{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9ef373e9",
   "metadata": {},
   "source": [
    "#### Loading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "88d29e84",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "reading tree\n",
      "\n",
      "importing quantifications\n",
      "\n",
      "reading in files with read_tsv\n",
      "\n",
      "1 \n",
      "2 \n",
      "3 \n",
      "4 \n",
      "5 \n",
      "6 \n",
      "7 \n",
      "8 \n",
      "9 \n",
      "10 \n",
      "11 \n",
      "12 \n",
      "\n",
      "\n",
      "found matching transcriptome:\n",
      "[ GENCODE - Mus musculus - release M25 ]\n",
      "\n",
      "loading existing TxDb created: 2023-02-17 18:26:37\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "loading existing transcript ranges created: 2023-02-17 18:26:39\n",
      "\n",
      "fetching genome info for GENCODE\n",
      "\n",
      "Missing txps 2380\n",
      "\n",
      "Aggregation Started\n",
      "\n",
      "Aggregation Ended\n",
      "\n",
      "Setting inf rep depth to mean depth\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress:   1 on 100Progress:   2 on 100Progress:   3 on 100Progress:   4 on 100Progress:   5 on 100Progress:   6 on 100Progress:   7 on 100Progress:   8 on 100Progress:   9 on 100Progress:  10 on 100Progress:  11 on 100Progress:  12 on 100Progress:  13 on 100Progress:  14 on 100Progress:  15 on 100Progress:  16 on 100Progress:  17 on 100Progress:  18 on 100Progress:  19 on 100Progress:  20 on 100Progress:  21 on 100Progress:  22 on 100Progress:  23 on 100Progress:  24 on 100Progress:  25 on 100Progress:  26 on 100Progress:  27 on 100Progress:  28 on 100Progress:  29 on 100Progress:  30 on 100Progress:  31 on 100Progress:  32 on 100Progress:  33 on 100Progress:  34 on 100Progress:  35 on 100Progress:  36 on 100Progress:  37 on 100Progress:  38 on 100Progress:  39 on 100Progress:  40 on 100Progress:  41 on 100Progress:  42 on 100Progress:  43 on 100Progress:  44 on 100Progress:  45 on 100Progress:  46 on 100Progress:  47 on 100Progress:  48 on 100Progress:  49 on 100Progress:  50 on 100Progress:  51 on 100Progress:  52 on 100Progress:  53 on 100Progress:  54 on 100Progress:  55 on 100Progress:  56 on 100Progress:  57 on 100Progress:  58 on 100Progress:  59 on 100Progress:  60 on 100Progress:  61 on 100Progress:  62 on 100Progress:  63 on 100Progress:  64 on 100Progress:  65 on 100Progress:  66 on 100Progress:  67 on 100Progress:  68 on 100Progress:  69 on 100Progress:  70 on 100Progress:  71 on 100Progress:  72 on 100Progress:  73 on 100Progress:  74 on 100Progress:  75 on 100Progress:  76 on 100Progress:  77 on 100Progress:  78 on 100Progress:  79 on 100Progress:  80 on 100Progress:  81 on 100Progress:  82 on 100Progress:  83 on 100Progress:  84 on 100Progress:  85 on 100Progress:  86 on 100Progress:  87 on 100Progress:  88 on 100Progress:  89 on 100Progress:  90 on 100Progress:  91 on 100Progress:  92 on 100Progress:  93 on 100Progress:  94 on 100Progress:  95 on 100Progress:  96 on 100Progress:  97 on 100Progress:  98 on 100Progress:  99 on 100Progress: 100 on 100  "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "suppressPackageStartupMessages(source(\"../tree_helper.R\"))\n",
    "suppressPackageStartupMessages(library(TreeSummarizedExperiment))\n",
    "suppressPackageStartupMessages(library(phangorn))\n",
    "\n",
    "metaData <- read.delim(\"/fs/cbcb-lab/rob/students/noor/Uncertainity/real_datasets/GSE100505_EDL_MAST/SRR_Acc_List.txt\")\n",
    "quantDir <- \"/fs/cbcb-lab/rob/students/noor/Uncertainity/real_datasets/GSE100505_EDL_MAST/sal_out/mode_gcbias=True/posttype=gibbs_npost=100_tf=100\"\n",
    "samples <- metaData$RUN\n",
    "files <- file.path(quantDir, samples, \"quant.sf\")\n",
    "colData <- cbind(data.frame(files = files, names = samples), condition = as.factor(metaData$TissueName))\n",
    "\n",
    "saveDir <- \"../environment/mouseMuscle\"\n",
    "clustFile <- \"/fs/cbcb-lab/rob/students/noor/Uncertainity/real_datasets/GSE100505_EDL_MAST/term_out/mode_gcbias=True/posttype=gibbs_npost=100_tf=100/no_threshold0/cluster_nwk.txt\"\n",
    "\n",
    "tseCons <- beaveR::buildTSE(treeTermFile = clustFile, coldata = colData)\n",
    "treeCons <- rowTree(tseCons)\n",
    "l <- length(treeCons$tip)\n",
    "\n",
    "yAll <- beaveR::computeSizeFactors(tseCons)\n",
    "yAll <- beaveR::scInfReps(yAll)\n",
    "yAll <- fishpond::labelKeep(yAll)\n",
    "metadata(yAll)[[\"infRepsScaled\"]] <- TRUE\n",
    "\n",
    "yTxps <- fishpond::swish(tseCons[1:l,],x=\"condition\")\n",
    "yInn <- fishpond::swish(tseCons[(l+1):nrow(yAll),],x=\"condition\")\n",
    "\n",
    "load(file.path(saveDir,  \"trenDiDegNodes.RData\"))\n",
    "load(file.path(saveDir,  \"detNodes.RData\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9928d16b",
   "metadata": {},
   "source": [
    "### Plotting the trenDi nodes that map to DEGs but dont contain a single DET\n",
    "We use plot gardener"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76f28cd7",
   "metadata": {},
   "source": [
    "#### Preparing data for plot gardener"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "11b85033",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(org.Mm.eg.db)\n",
    "library(TxDb.Mmusculus.UCSC.mm10.knownGene) ## this one corresponds to the version we need\n",
    "suppressPackageStartupMessages(library(plotgardener))\n",
    "suppressPackageStartupMessages(library(plyranges))\n",
    "suppressPackageStartupMessages(library(ggtree))\n",
    "suppressPackageStartupMessages(library(\"grid\"))\n",
    "suppressPackageStartupMessages(library(\"ggplotify\"))\n",
    "suppressPackageStartupMessages(library(\"ggplot2\"))\n",
    "suppressPackageStartupMessages(library(\"ggpubr\"))\n",
    "\n",
    "txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene\n",
    "txpMouse <- transcripts(txdb)\n",
    "\n",
    "txpMouse <- txpMouse %>%\n",
    "  mutate(tx_id = as.character(tx_id))\n",
    "\n",
    "txpMouse <- txpMouse %>%\n",
    "  mutate(gene_id = mapIds(\n",
    "           txdb, keys=tx_id,\n",
    "           column=\"GENEID\", keytype=\"TXID\")\n",
    "         ) %>%\n",
    "  filter(!is.na(gene_id))\n",
    "txpMouse <- txpMouse %>%\n",
    "  mutate(ensID = mapIds(\n",
    "           org.Mm.eg.db, keys=gene_id,\n",
    "           column=\"ENSEMBL\", keytype=\"ENTREZID\")\n",
    "         ) %>%\n",
    "  filter(!is.na(ensID))\n",
    "txpMouse <- txpMouse %>%\n",
    "                filter(tx_name %in% intersect(treeCons$tip, mcols(txpMouse)[[\"tx_name\"]]))\n",
    "\n",
    "\n",
    "genesMouse <- genes(txdb)\n",
    "genesMouse <- genesMouse %>%\n",
    "  mutate(ensID = mapIds(org.Mm.eg.db, \n",
    "                            gene_id, \"ENSEMBL\", \"ENTREZID\")) %>% \n",
    "    filter(!is.na(ensID))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9d440cb",
   "metadata": {},
   "source": [
    "#### Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a42390d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "trenDiNodes <- detNodes[[\"trenDi\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c8363171",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ENSMUST00000120135.7\" \"ENSMUST00000055087.6\"\n",
      "[1] \"Genes ENSMUSG00000021457\"\n",
      "              tx_name1               tx_name2               tx_name3 \n",
      "\"ENSMUST00000120135.7\" \"ENSMUST00000118756.7\" \"ENSMUST00000150672.1\" \n",
      "              tx_name4               tx_name5 \n",
      "\"ENSMUST00000055087.6\" \"ENSMUST00000140339.1\" \n",
      "[1] \"ENSMUST00000120135.7\" \"ENSMUST00000055087.6\" \"ENSMUST00000118756.7\"\n",
      "[4] \"ENSMUST00000140339.1\" \"ENSMUST00000150672.1\"\n",
      "[1] TRUE\n",
      "[1] TRUE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "legend[legend1]\n",
      "\n",
      "text[text1]\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in plotTranscripts(params = par, x = 1.5, y = 5.5, width = 5, height = 8.5, : unused argument (transcriptHighlights = hilite)\n",
     "output_type": "error",
     "traceback": [
      "Error in plotTranscripts(params = par, x = 1.5, y = 5.5, width = 5, height = 8.5, : unused argument (transcriptHighlights = hilite)\nTraceback:\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABpAAAAc4CAIAAACweRKhAAAACXBIWXMAABJ0AAASdAHeZh94\nAAAgAElEQVR4nOzde5QU1aEv4N3DPBgewxsUUCMQEIxIrosIKggxAioaVAgauBo1SgLxJIhL\nMSdcTU7EV8SDImoIiSjxgWhAoyKooIjPaERQDioqoojIIMhLeUzfP2qlT2e6Z5jpGaDQ71v+\nMezatav23tW9Vv/cVZVIJpMBAAAAAIiHvH19AgAAAADA/xLYAQAAAECMCOwAAAAAIEYEdgAA\nAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwA\nAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHY\nAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEi\nsAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABi\nRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAA\nxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAA\nAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAA\nAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwA\nAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHY\nAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEi\nsAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABi\nRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAA\nxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAA\nAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAA\nAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwA\nAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHY\nAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEi\nsAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABi\nRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAA\nxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAA\nAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAA\nAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwA\nAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHY\nAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEi\nsAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABi\nRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAA\nxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAA\nAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAA\nAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwA\nAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHY\nAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEi\nsAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABi\nRGAHAAAAADEisAMAAACAGPnGBXa33nprIpFo2LDh66+/vq/P5WsukUgkEonhw4fv6xP5pjMR\nAAAAsH/5xgV2o0aN+ulPf7p58+aBAweuXbt2X58OAAAAAPybr1Vg169fv2gxUceOHZPJZEXV\nJk6c+O1vf/vjjz8+77zzqtjyzJkzE9VkQdOVV1555ZVXnnHGGTVv6qabbrr++utr3s43Uy1O\nRC0yp7XohRdeuOSSS4499tgDDjiguLg4Pz+/UaNGhx122JlnnvnHP/5x8+bN6ZXXrl3bokWL\nRCLRqFGjVatWVd7yxx9/3Lhx40Qi0bJly9T/4Rg7dmz0Lffkk09WtOOsWbMKCwsTiUS7du1W\nrlxZ8z4CAAB8oyQqCbb2L2+//fZhhx2Wl5dXWFi4bdu2OXPm9O/fv6LK8+fP//73vx9CmDZt\n2jnnnLPbxmfOnDlkyJBqnc+wYcOmT59erV3IauPGjU2bNm3SpMm6dev29blQO8xpbSktLT3n\nnHMee+yxSuq0bNnynnvuOeGEE1IlqS+0/v37z5kzp5J9Bw4c+Oijj4YQHnzwwVTmO3bs2Ouu\nuy6EMG/evB/84AeZez388MODBw/esWNHu3bt5s+ff/DBB1e/ZwAAAN9o+fv6BGrNpEmTkslk\n7969mzZt+uCDD06ePLmSwK5v374DBw78+9//fsUVVwwePLhevXqVN967d+958+aVK1y6dOno\n0aNDCEOGDLnooovKbT3wwANz6gflvfTSS2VlZfv6LKhN5rRWbN++vV+/fq+99loI4aijjvrp\nT3961FFHtWzZMplMrl27dsmSJXfdddezzz67du3aU0899cUXX+zatWu04+DBg88+++x77733\niSee+Mtf/lLRWuO77747SuuGDRtW9RWajzzyyJAhQ3bs2NG+ffv58+cfdNBBtdFXAACAb5av\nSWC3efPmadOmhRCGDh3avHnzBx988NFHH/3www8rWdlxxRVX/P3vf1+9evWdd945cuTIyttv\n2bJl5kKSunXrRn8cfPDBWZeZUCteeOGFfX0K1DJzWitmzJgRpXXnnHPOX/7yl7y8/33Ewbe+\n9a3vfe97F1xwwW9+85urr75627ZtY8aMSf+/DpMmTVqwYMEnn3xyySWX9O/fv3Xr1uUa//TT\nT3/1q1+FENq0aXPLLbdU8ZQeffTRwYMHb9++/dvf/vb8+fPbtGlT004CAAB8I31NnmF31113\nffHFF0VFRT/60Y8GDhzYuHHjXbt23XHHHZXscswxx3z3u98NIdx888177sTy8/MTicRVV11V\nVlY2fvz49u3b5+fnX3XVVel15s6dO3z48I4dO9avX7+goKB58+a9e/eeMGHC1q1bMxusW7du\nIpEYN25cCOGtt9664IILDj300Lp165aUlPTo0WPSpElZFy4tXbp01KhRXbt2LSkpyc/Pb9iw\n4RFHHDFy5Mg33ngj62mvWLHi4osv7tKlS8OGDevWrduhQ4cRI0YsW7asWr3LfJbftm3bosIr\nrrgihLBgwYIzzjijbdu2RUVFLVq0OOmkk2bPnp3e/qWXXhq1H0IoLS2N9u3QoUPO/ap8PBcu\nXDhw4MBWrVoVFxd36NBhzJgxpaWlUbV//OMfQ4cObdu2bWFhYevWrYcPH/7uu+9mbbDqs1nd\n0chtl1DBW2Jzu5B27Nhxyy239OzZs0mTJvXr1z/88MN/+9vfbtiwIYRwxBFHJBKJk08+ufIB\n3+2chhDef//9MWPGdOvWrUmTJoWFhS1btuzVq9fVV1/9+eefV954rZzt2rVrx40b171796ZN\nmxYWFrZq1apPnz433HDDpk2bMivnNozVPUpFnn322eiPcePGpad16X73u99997vf7datW8uW\nLb/88stUedOmTadMmRJC2LBhw89+9rPMHUeOHLl+/foQwp/+9KcmTZpU5Xwee+yxM888c/v2\n7R07dlywYIG0DgAAIHfJr4XDDz88hDB06NDon9Hvz5YtW3711VeV7DV+/PhoEBYvXpzDQRcu\nXBjtPmbMmIrq1K9fP4Rw2WWXXX311akxv/LKK6Otu3btquTdFB07dly9enW5Bhs1ahRCGD16\n9Jw5c6LGyzn77LPL7XLbbbfVqVMnVSH9h32dOnVuvfXWcvWnTp1aUFCQ2XJBQcEdd9xR9d5F\n/xw2bFj6LlHhyJEjb7/99qwRw+jRo1OVf/vb37Zq1aqwsDCEkEgkWrVq1apVqx49euTWr6xS\n43nffffl55dfcNqtW7dt27Y98MAD0Tmka9as2UcffZTeVA6zWa3RqOEu5SYihwtp8+bNxx13\nXKpC9EqBEMK3v/3t999/P7rz8cwzz6x8wCuf02Qyedttt6Uuv7y8vJKSkugoIYQmTZosWLCg\n8vZreLazZ89u0KBBtEt+fn76zfIHHnjgyy+/XK5+DsOYw1EqkrrePv744yruUs75558ftTB9\n+vT08vvvvz8qv+iiizL3uvzyy6Ot8+bNSxU+/vjjRUVFIYTDDjss81IHAACgWr4Ogd3TTz8d\n/XqcO3duVPLqq69GJX/9618r2fH111+Pql177bU5HLcqgV30e/6CCy5o3Lhx586dx48fP3ny\n5GeeeSbaeuONN0Yt9OnT57nnntu4ceNXX321fPny1C26/fv3L9dg06ZNQwg//OEPmzZtOmDA\ngOeff37Tpk2lpaUPP/xwaj3LwoULU/XfeeedKIc68sgj586du2HDhmQy+fnnn8+ePfuII46I\nMpElS5ak6j/xxBNRI926dXvkkUfWrFmzcuXK++6775BDDokSlieffLKKvYvaKZcTRRlTz549\ni4qKTj311EWLFn3++eelpaX33HNPq1atol0efvjh9F2iZxE2a9YsvbC6/apINJ6nnXZaSUnJ\nxRdf/O67727duvX111/v3bt3dDJXXXVVw4YNhw4dumTJkq1bt77zzjtnn312tOmXv/xlelM5\nzGYOo5HDLlknoroXUjKZvPjii6PyE088cenSpclkctOmTVOmTCkpKenVq1fz5s1DWmheuaxz\nmkwmZ82aFR2iY8eOjz766Pbt26OjTJ06tXHjxiGEhg0bfvDBB1U5RA5nu2jRoigC7tat25NP\nPrljx45kMrlmzZpJkyZFS8yaN2/+ySef1HAYczhKRa655proEP/v//2/qtTPtHHjxui5Ac2a\nNVuzZk1U+Nlnn7Vs2TKEcOihh27atClzr8zAbs6cOVFa17lz5yqePAAAAJX4OgR20dPQv/Wt\nb5WVlaUKo9tdjz322Ep23LlzZ7TO5Yc//GEOx61KYBf9nm/WrFn37t23bt1abmv79u1DCA0a\nNMj8VTxo0KCo8Q8//DC9vFmzZlH5aaedtmvXrvRNDz74YLRp7NixqcIJEyZEhW+88Ua5Q6xf\nv76kpKRRo0Y33XRTVFJWVtaxY8cQwkEHHbRx48b0ysuXL49+kKcvhqq8d1lzotSauAEDBpQ7\n/5deeinadNxxx6WXZw13qtWvSqTG8xe/+EV6+Ycffpha2DVw4MD0S+vLL7+M9jr88MPTd8lh\nNnMYjRx2yToR1b2Q1q9fHy1869Sp07Zt29LrP/PMM6m1fjUJ7Hbt2hXlwo0bN85copV6melP\nfvKT3baf29lGUW+nTp2++OKLcg2+8sor0V6jRo1KL6/uMOZ2lIqsWbMmyjFDCGeddVb0Ko+q\n7JjuySefjC711HrDs846K4SQl5eXCt/LKRfYzZ07N3qg5+GHH/7pp59W9wQAAADItN8HdqtW\nrYpWWv3ud79LL588eXL0k7Ly212PPvroEEK7du1yOHRVArvU7/n58+dnbl2xYsWLL774/PPP\nZ266/fbbox0fe+yxzAYTicQ777xTbpfU06/Sb/RLPVGu3P2bkXIpW+qRWNdcc01m5Z///Od9\n+vQZPHjwl19+WZXeZc2JUnnTP/7xj8xdjj322Kh3paWlqcKs4U61+lWJqAt16tTJDBqiAC6E\n8OKLL5bbdMopp4QQ6tWrl16Yw2zmMBo57FJJYFf1C+mee+6JCidOnJh53B//+Mc1D+zmzp0b\nNXL55Zdn3atbt27RsFd+q3tuZ7to0aKocNq0aVnbHDhwYAihUaNG6cFcdYcxt6NUYuHChS1a\ntAj/0rx580GDBl1zzTXPPPNM1T8Fo0aNinafMWNG6jGImfdWp6QHdvPmzSsuLg4htGjRYu3a\ntVU8IgAAAJXb7186cdttt+3cuTMvL++8885LLx82bFj0ZKhUcpfVt771rRDCqlWrKno2fK1o\n1KhR6hbLdO3atTv66KN79uyZuSlavBZC+OKLLzK3dujQodxz+kMIDRo0iJ6ilf7c+q5du0Z/\njBgx4pNPPim3S/RLOyUVQR5//PGZB508efL8+fMfeOCBaKldSkW9q0Tz5s2POuqozPIob0om\nk0uXLq28hWr1a7e6dOkS3QOYLiopLi7u3r171k1bt25Nv2xyns0cRqPmAxip+oWUun8861xH\na7JqKBVm9evXL2uFvn37hhC2bt26ZMmSypvK4WwXLFgQ/ZH+5Lt0UVMbN278n//5n3Kbqj6M\nNTlKVscdd9yyZct+/etfRzdEr1u3btasWVdcccXxxx/fuHHjvn373njjjal3p1Tkuuuui85/\n1KhR0QNADzvssNQjPivx9NNPn3baadu2bQshfPbZZ/fee29VzhkAAIDdKv+U/f3L9u3b//Sn\nP4UQBgwY0LZt2/RNJSUlQ4YMmTZt2l//+tfrr7++pKQkawvRo6x27NjxxRdfpG4uq3UdO3as\n6B2OO3bsePzxx5966qnly5dv2LAhWjoUQti4cWNUIfmvFVLpokfmZ6pbt+6WLVt27dqVKjnt\ntNP69OmzYMGCRx999JBDDunVq9eJJ57Yp0+f7t27p7+xIfL2229Hfxx66KG10ruKdOnSJWt5\nlJ+GED7++OPKW6hWv3brwAMPzCyMXozQokWLzN6l3kFRVlaWvjW32cxhNGo+gJGqX0gffPBB\n9Ee7du0y62eNKavrnXfeif6I7svOlFrz+N5772XNK1NyONvly5eXO0pFVq5cWW78qz6MNTlK\nRZo1a3b11Vf/13/916uvvrpgwYIXX3zxxRdfXL169fbt2xcsWLBgwYJx48b9+te//vWvf13R\n57R+/fp33nln7969P/vssxBCfn7+XXfdFd3lWrnoIXpnnHHGm2++uXz58tGjR7dt2zZ6RgEA\nAAA1sX8HdjNmzFi7dm0I4fvf/37mgpQ+ffpMmzZt8+bN06ZNSz2BvpzUix23bNmy5wK76OUM\nmZYsWTJkyJDUb/iqy/oW16zq1Knz2GOP/eY3v7ntttu2bdv29NNPR+/oaNy48cCBA0eNGtWj\nR49U5Q0bNkR/pF5hWRUV9a4SFQ11w4YNoz/SFyVlVa1+7VYl41n1oc55NnMYjZoPYKTqvdu8\neXO5Q6Rr1qxZXl5eDZeppi6/inqXit1TAWhFcjjb1NF3K/PoVR/Gmhylcnl5ed27d0+tBv3w\nww+fffbZWbNmzZ49e9u2bePGjXv//fenTp1a0e7HHnvsyJEjJ02aFEIYPXp05qrSrIqKim68\n8cZRo0atWLGiR48e69atGzZs2FNPPXXMMcdU6+QBAAAoZ/++JTb6eRlCuPTSSztnSN0ke9tt\nt1XUQmrFU+oNA3tC1jVfGzdu7NevX5TvnH/++c8+++y6deuiV0Ymk8kHHnigto5eXFx84403\nrlmzZvr06cOGDYtWk23YsGH69Ok9e/b81a9+lblL1oVgFclhRVtqhVo5qQylKkv2cujXnlOT\n2cxhNGplAKslajmRSFT0San5EVMtV3T5Vf3TmsPZpko+/fTTHZUaOnRo1TpU2XH36FFCCAcf\nfPDw4cNnzpy5ePHiaN3ln//859RDKrM6/PDDoz+quLIvhHDXXXdFz79r37793/72t6Kioi+/\n/PK0005LrdUFAAAgN/txYPfqq6+mXotZuWXLls2fPz/rpq1bt0Z/RA+825vuvPPONWvWhBAu\nv/zyqVOn9urVq1mzZtELNEII0WOhalFJScmwYcOmT5++evXqt9566/e//330vPyJEyfefffd\nUZ3UyqaqrwPKTdZHuYXdLYzKqir92gtqMps5jEYtDmAVRR+QZDKZ+sik27Bhw86dO2t4iN1e\nfqlFZ7tdDJvD2aba3LhxY36lahLu752jpOvSpcsf/vCH6O+HH364VtpMST2cMYRw3HHHRSv4\nSktLTzrppGjtMwAAALnZjwO7W265JYSQSCRWrFhR0Ts11qxZE92tVtGrJ6JnNhUUFFT0kLs9\nJ5U2/sd//Efm1jfffHPPHbpz587/+Z//uXjx4uj39s033xyVd+rUKfqjig+8z9mKFSuylq9c\nuTL6o1oP0UupqF97QU1mM4fR2EMDWInWrVtHf3z44YeZW1955ZWaH+Kwww6L/qjo8ks95C51\noVYkh7NNLStbvHhxFU42R7V4lM2bN7/wwgu33377bqOx1HDt6SB+2LBh0eub33vvvYEDB2ZN\nSwEAAKiK/TWwKy0tvf/++0MIJ5xwQtbnykdatWp16qmnhhBmzZq1evXqzApRwHHwwQfX+i2E\nu5VaDNWiRYtym7Zs2TJt2rRaOUppaelTTz2V+R7VEEKbNm2il4qmRib15srHH388s/6IESPy\n8/MLCgqipWQ1sWLFivfffz+z/Pnnnw8hFBQUpO7Oq0i1+rUX1GQ2cxiNmg9gdaUazLqsdcaM\nGTU/ROqNrnPmzMlaYe7cuSGEJk2adO7cufKmcjjb1NErun/5ySefnDt37vbt2ys/dOVq8Si3\n3nrrMccc8/Of/zz6XxeVSL0zt6KXY9SiK6+8cvjw4SGEV1555ayzzkp/4QYAAABVt78Gdn/6\n05++/PLLEMKFF15Yec2f/vSnIYSdO3f+8Y9/LLdp165d0dKn73znO3vmNCuTWgS0ZMmScmd1\n4YUXpm6irOLbA7LavHlzmzZtfvCDH1xxxRWZW3fs2PHWW2+FEA455JCo5Ljjjote0DllypRy\nedCqVatmzpy5a9euI4888oADDsj5lFJ+//vflyv55z//uWjRohDCiSeemHoZSPjX89q++OKL\n1APaqtuvvaCGs1n10ajJLjXRv3//6I8pU6aUe8bcSy+9dOedd1artcw5DSH06tUrdfl99NFH\n5Xa5//77o0cEnnfeebt9bGIOZ3v00Ud37do1hDBz5szMZ7198skn5557bv/+/c8666zdd69i\ntXiUn/zkJ9GNz9dee20ld38vW7Zs7NixIYREInHmmWfW5OSraOrUqVH0/8gjj/ziF7/YC0cE\nAAD4+tkvA7uysrLbb789hNCiRYtBgwZVXrl///4HH3xwCGHKlCnlHly1dOnSaGFUz54999jJ\nVuiUU06J/vjZz34W3eu3c+fOp59+um/fvg899ND06dOjrbNnz855lUqDBg2iQHPatGlDhgyZ\nP3/+Z599tmvXrk2bNj3//PODBg2K7qyMMs0QQiKRiFbrbN68uXfv3tOnT//oo48++OCDe++9\nt2/fvuvXr08kEuPHj69Zv0MI4cgjj5w2bdrPfvaz9957L4Swa9euOXPmDBo0KMpWysVw0bKg\nHTt23HDDDR9//PGSJUsKCgqq1a+9oCazWa3RyHmXGjr00EMHDhwYQli0aNG5554b3Wq6bdu2\nu+6666STTjr99NOrFRFmzulXX32VSCQmTZqUSCQ2bdr0/e9///HHH9+xY0cI4bPPPrvpppui\nd8i0bdv2N7/5zR4628mTJ9epU6esrOzUU0+95ZZbPvvss2QyuXbt2rvvvvvYY49dvXp1fn5+\ndMtnTdTWUVq1anXHHXckEomdO3eec845PXr0uP766+fMmfPaa68tXrz4ueee+8tf/jJs2LBu\n3bp9/PHHIYSxY8funf8zUVhYOGvWrA4dOoQQbr/99muvvXYvHBQAAODrpqKnv8XZrFmzopO/\n9NJLq1I/9ev3gQceSC9PZU9LlizJ4TQWLlwY7T5mzJiK6kQvQOjfv3/mpugXe2oiGjZsGK0b\nKioqmjlzZllZWepxV8XFxUOGDNltg6mtJ5xwQqpk27ZtAwYMqOQCGD16dLlGpkyZEj34r5yC\ngoI77rijir1L/mtZ07Bhw9ILoz7++Mc//sMf/hA9Vr+oqCh9wdTVV19drp3MF4asWrUqh35V\nMmKnnHJK5qbjjz8+hNC+ffvMTSNGjIiOknoPbG6zmcNo5LBL1omo7oWUTCZXrVoVZd+RBg0a\nRDeSH3300evWrYsisKFDh1Y41mmyzmm0adq0aUVFRVFhXl5ecXFxqk779u2XLVtWlfZzPtuH\nHnooPctLv1O+pKTkiSeeyDpQ1RrGHI5SiUcffXS3N7o2atRo4sSJu20q9Tbtv/zlL5XXvPzy\ny6Oa8+bNq6jO8uXLo0dJJhKJv/71r1XvEQAAAMlkcr9cYXfrrbdGf1RxCdX55wIPzBQAACAA\nSURBVJ8f/SQu9+qJ6DFSXbp02Se3xCYSiQcffPC666478sgji4uLy8rKOnTo8Mtf/vKNN944\n88wzo1+5PXr0KC4ubtiwYc6PJKtbt+5jjz320EMPDR06tEOHDvXq1cvLy2vYsOF3vvOdESNG\nvPzyyxMmTCi3y09/+tM333xz1KhRnTp1ql+/fmFhYbt27S688MLXX3/9oosuqnG/QwihrKxs\nzJgxTz311BlnnNG8efM6deq0aNHilFNOmTdv3q9//etylfv06TNlypROnToVFhY2adKkR48e\n9erVy6Ffe1RNZrNao5HzLjXXtm3bf/7zn5dddlmnTp2Ki4vz8/OPOuqoiRMnPvPMM82aNYtu\nbk29GLdyWec02nTOOecsX758zJgxXbt2bdCgwc6dO1u2bHniiSfeeuutb775ZurFFHvobE8/\n/fQVK1aMGzeue/fuTZs2rVOnTuvWrY8++ujx48cvX768X79+VR2sStXiUU4++eQVK1bMmDHj\nvPPO+z//5/80a9assLAwPz+/UaNGXbp0+dGPfvSnP/3pgw8+yPoulD2qY8eODz30UGFhYTKZ\nPO+88yp6TzcAAABZJZL//oCnb44XXnjhmGOOCSHcdtttP/vZz/b16XxT5Ofn79q1a+jQoffd\nd9++Ppd9L4fRiOcAfvXVV3Xr1g0h/OIXv9jtOxD2uf3rbAEAAPgG2i9X2NWK6MlKbdq0Offc\nc/f1ucD+4fPPP898HUQI4ZVXXon+2O37W/em/etsAQAAIOUbGtgtWLDg4YcfDiFcd9116Q/J\nArKKbvJt2rTpcccdl3rpbcoNN9wQQsjLy6v8wYJ7zf51tgAAAFDONzGw27p1a/QstlNOOWXY\nsGH7+nRgP5CXlzd06NAQwsqVK0888cQFCxZs3rx5586dixcv/tGPfhTF3+eee267du329ZmG\nsL+dLQAAAJRTpSfEf82MHj36nXfeadOmzZ///Od9fS6w37jqqquWLVv2t7/9bdGiRX379i23\ntX///rF6Htz+dbYAAACQ7hu3wm7y5Ml//OMf69ev//e//71ly5b7+nRgv1FQUPDQQw/NmjXr\n9NNPP+iggwoLC4uKitq2bTto0KCZM2c+9thj9evX39fn+L/2r7MFAACAdN/ct8QCAAAAQAx9\n41bYAQAAAECcCewAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAA\nADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAA\nAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0A\nAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7\nAAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYE\ndgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECM\nCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACA\nGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAAAABiRGAHAAAA\nADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0AAAAAxIjADgAA\nAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7AAAAAIgRgR0A\nAAAAxIjADgAAAABiRGAHAAAAADEisAMAAACAGBHYAQAAAECMCOwAAAAAIEYEdgAAAAAQIwI7\n4iuRSCQSieHDh+/rE6HW1K1bN5FInHXWWVWpnJ+fX/XKe0Hz5s0TicSAAQP29YnsH7LOdQ4f\n6mpdMxguAAD4ehDYAQAAAECM5O/rE/immDlz5pAhQ6q1y7Bhw6ZPn76Hzme/cOWVV4YQunbt\nWvOmbrrpph07dlx22WU1b4pvAhfMHlKLH+ocmFYAAGB/IbAjvq666qpaaWfjxo2XXnppkyZN\n/FCnKlwwe05tfahzYFoBAID9iMDu3yUS1aicTFa9bu/evefNm1eucOnSpaNHjw4hDBky5KKL\nLiq39cADD6zGyVCxl156qaysbF+fBfsNF8zXkmkFAAD2IwK7vaRly5Y/+MEPyhXWrVs3+uPg\ngw/O3EpteeGFF/b1KbA/ccF8LZlWAABgP+KlE3EXvSjzqquuKisrGz9+fPv27fPz88vdVjZ3\n7tzhw4d37Nixfv36BQUFzZs3792794QJE7Zu3ZrZYPQOwXHjxoUQ3nrrrQsuuODQQw+tW7du\nSUlJjx49Jk2alHURytKlS0eNGtW1a9eSkpL8/PyGDRseccQRI0eOfOONN7Ke9ooVKy6++OIu\nXbo0bNiwbt26HTp0GDFixLJly6rVu8wXSm7bti0qvOKKK0IICxYsOOOMM9q2bVtUVNSiRYuT\nTjpp9uzZ6e1feumlUfshhNLS0mjfDh065NyvilSxv5H3339/zJgx3bp1a9KkSWFhYcuWLXv1\n6nX11Vd//vnn6dVOOumkRCJRp06dNWvWZG3n008/jQZw4MCB6eVr164dN25c9+7dmzZtWlhY\n2KpVqz59+txwww2bNm3KbKQqF1i1Ggwh7Ny5c+LEid/73vcaNWrUoEGDTp06XXLJJR999FHF\n41eZRCIRQnj22WeHDBly0EEHFRUVtWzZcsCAAeXmun///olEoqCgoKLh2rRpU3FxcSKR+P73\nv1/RsXZ7wYQQCgoKQgjLly+/6KKL2rVrl/rs3HrrrRUt4KruAFaiWlfaXvhmqO5cV/SW2Nyu\nmSp2sCrTWitzlNsYhip/J6TkNly1eB0CAAB7XJJ0IVTjvxpbuHBhNAtjxoypqE79+vVDCJdd\ndtnVV1+dmrUrr7wy2rpr167Mn74pHTt2XL16dbkGGzVqFEIYPXr0nDlzosbLOfvss8vtcttt\nt9WpUydVIS/vf3PeOnXq3HrrreXqT506NQo1yikoKLjjjjuq3rvon8OGDUvfJSocOXLk7bff\nnn4mKaNHj05V/u1vf9uqVavCwsIQQiKRaNWqVatWrXr06JFbvypS9f5GB01VzsvLKykpSfzr\nRuwmTZosWLAgVTP1ypFbbrkl63FvvvnmqMKMGTNShbNnz27QoEFUnp+fX69evdT5HHjggS+/\n/HK5Riqfghwa3Lp1a+/evVN1oiQlhNC8efPXXnutuLg4hDB06NCqDGw0O8OHD588efJu53rm\nzJlR4bXXXpu1tbvuuiuqMH369IqOWPkF06xZsxDC4MGD582b17Bhw8zzyfzs5DCAlaj6lbZ3\nvhlymOuoZrkPdQ7tVKuDlU9rLc5RDmOYrM53Qs7DVYt9BAAA9g6B3b+LX2AX/QK84IILGjdu\n3Llz5/Hjx0+ePPmZZ56Jtt54441RC3369Hnuuec2btz41VdfLV++fOTIkVF5//79yzXYtGnT\nEMIPf/jDpk2bDhgw4Pnnn9+0aVNpaenDDz/cpk2baK+FCxem6r/zzjv5+fkhhCOPPHLu3Lkb\nNmxIJpOff/757NmzjzjiiOgX5pIlS1L1n3jiiaiRbt26PfLII2vWrFm5cuV99913yCGHRL+W\nn3zyySr2Lutv+yi46dmzZ1FR0amnnrpo0aLPP/+8tLT0nnvuadWqVbTLww8/nL5L//79QwjN\nmjVLL6xuvypSrf7OmjUrFSg8+uij27dvTyaTmzZtmjp1auPGjUMIDRs2/OCDD6LKW7ZsiX5g\n9+7dO+uhjznmmBBC48aNv/zyy6hk0aJFUcjVrVu3J598cseOHclkcs2aNZMmTWrSpEn0k/6T\nTz5Jb6TyKcihwUsuuSTq46mnnvruu+8mk8mysrJFixZ973vf69SpUw6BXc+ePQsLC6O53rhx\n47p16+69994DDjggOsojjzwSVd6+fXtU2KlTp6ytnXTSSdFwbd26tfLjZr1gkv8K7Pr169es\nWbN+/fotWrRo06ZN69at+9vf/ta6devMz05uA1iRal1pe+GbIZnTXGf9UOfQTg4drGhaa3GO\nchjDan0n5DxctdhHAABg7xDY/bv4BXbRL8BmzZp17949M2ho3759CKFBgwabNm0qt2nQoEFR\n4x9++GF6eRQ6hBBOO+20Xbt2pW968MEHo01jx45NFU6YMCEqfOONN8odYv369SUlJY0aNbrp\nppuikrKyso4dO4YQDjrooI0bN6ZXXr58eVFRUQghfWFL5b3L+ts+tSZuwIAB5c7/pZdeijYd\nd9xx6eVZf6hXq18VqVZ/d+3aFWUrjRs3zlzfNGfOnOh8fvKTn6QKozVEeXl5mfU//PDDaBnO\niBEjUoVR1NipU6cvvviiXP1XXnklyjpHjRqVXl75FFS3wQ0bNkS97ty581dffZVef9OmTd/6\n1reiPlYrsAshDBw4sKysLH3Tyy+/HG3q1atXqnDs2LFR4XPPPVeuqXXr1kWrkEaOHLnb41Ye\n2GX97KRil/TPTjKnGcmqup+svfDNkNtcZ36oc2snhw5WNK21NUfJ6o9hDt8JuQ1XLfYRAADY\nOwR2/y5+gV3qF+D8+fMzt65YseLFF198/vnnMzfdfvvt0Y6PPfZYZoOJROKdd94pt0vqSUZn\nnnlmqjD1OLOPPvoo8yjlIp5nn302qnzNNddkVv75z3/ep0+fwYMHp1aEVd67zN/2ybQQ5x//\n+EfmLscee2zUu9LS0lRh1h/q1epXRarV37lz50aVL7/88qytdevWLYRQr1691O/w1C/2zLti\nr7/++mjTokWLopJFixZFJdOmTcvafvSou0aNGqVHCZVMQQ4N3nfffdEuEyZMyKx/0003ZU0T\nKpKa69dffz1zazTXeXl569evj0pWrFgRhZjnn39+ucqpj0PWy6acygO7RCLx3nvvldu0ZcuW\n6NDpn53cZiSr6n6y9sI3Q25znfmhzq2dHDqYdVprcY6S1R/DHL4Tchiu2u0jAACwd3jpxP6h\nUaNG6Q8tSmnXrt3RRx/ds2fPzE3RyqkQwhdffJG5tUOHDuWeuR5CaNCgQfTcpfRnkHft2jX6\nY8SIEZ988km5XaLbr1JSEeTxxx+fedDJkyfPnz//gQceiFaIpFTUu0o0b978qKOOyiyPQpxk\nMrl06dLKW6hWvypSrf6mfjb369cva2t9+/YNIWzdunXJkiVRyQ9+8IPoNs8ZM2aUqxz9bu/Q\noUN0Y2wIYcGCBdEfxx13XNb2o0HeuHHj//zP/5TblHUKcmjwn//8Z/RH6qzS5fYq5AMOOODI\nI4/MLI/muqysLDXX7dq1iw4xY8aMLVu2pFe+9957QwhHHnlk1sumWg477LBDDz20XGG9evWi\nh4Klf3ZqMiPlVPeTtRe+GWprrnNrJ+cOllOLc5RS9THM4Tshh+HaE30EAAD2tPx9fQJUSceO\nHbM+dD+EsGPHjscff/ypp55avnz5hg0booUYIYSNGzdGFZL/WtWS7qCDDsraWt26dbds2bJr\n165UyWmnndanT58FCxY8+uijhxxySK9evU488cQ+ffp07949/Y0Nkbfffjv6IzPRyK13FenS\npUvW8tQdYR9//HHlLVSrXxWpVn/feeed6I/o3sZM0V1+IYT33nsvypXq1Klz1lln/fd///ei\nRYtWr16delDau++++9prr4UQzjnnnNTuy5cvL9dORVauXFluALNOQQ4NfvDBB1FJdKNfRR2s\nlsMOOyxreWquV61alSq86KKL5s2bt3nz5hkzZpx33nlR4erVq6PA64ILLsjhBMpJzUI5mZ+d\nmsxIOTl8svb0N0NtzXXO7eTWwXJqcY5Sqj6GOXwn5DBce6KPAADAniaw2z9EbwbItGTJkiFD\nhqR+j1Vd1ndNZlWnTp3HHnvsN7/5zW233bZt27ann3766aefDiE0btx44MCBo0aN6tGjR6ry\nhg0boj9SryOsiop6V4noceyZUu/uTF/GklW1+lWRavU3Vbmiky8pKYn+SCUOIYThw4f/93//\nd1lZ2YMPPnjxxRdHhdF6sUQi8X//7//NbH+30tuPZJ2CHBpMrWnKOiDFxcV16tRJDyyqInoo\nfqbUITZv3pwq/OEPf9iqVatPP/30z3/+cyqwu//++8vKyoqKioYNG1atQ2cVvaukKmoyIxU1\nVcVP1l74Zqituc6tnZw7WE4tzlFK1ccwh++EHIZrT/QRAADY09wSu3/IuuZr48aN/fr1i36y\nnn/++c8+++y6deui1/8lk8kHHnigto5eXFx84403rlmzZvr06cOGDTvwwANDCBs2bJg+fXrP\nnj1/9atfZe5SlbUtKVVf0ZZSWFiYtbysrCz6oypL9nLoV0Wq0t/oGWeVVE6Vp2qGEI466qjO\nnTuHf78r9v777w8h9O7dO7XKLKR1+dNPP91RqaFDh5Y7dNYpyKHByschmUymJqjqKro8UqOU\nPtcFBQU/+clPQgjPPfdcavlSlG8OGjQodbPk3lGTGalIVa60vfPNUFtznUM7tdjBPTFHVZfD\nd0IOw7Vv+wgAAORGYLcfu/POO9esWRNCuPzyy6dOndqrV69mzZqllv9s27atdg9XUlIybNiw\n6dOnr169+q233vr9738fPWF94sSJd999d1QntU6k6ms6clPR06lSi61SS+12qyr9qki1+rvb\nyqnlLeWW20TrwqK7YkMIS5YsefPNN0MI5557btb2N27cmF+p9ECwKidc9QZTq36yrnD8/PPP\nq5XkRiqa61R5ubm+8MILo/OJpu+999575ZVXQi3dD1sttTgj1brS9s43Q23NdQ7t1GIHa/1T\nUy05fCfkMFz7to8AAEBuBHb7sZdeein64z/+4z8yt0aZzh7SuXPn//zP/1y8eHG0ZOnmm2+O\nyjt16hT9sacfXr5ixYqs5StXroz+qNZD9FIq6ldFqtXf1LPYKqqcWhGWajYyfPjwRCKRTCYf\nfPDB8K/ldfXq1Rs8eHB6tdTDpxYvXrzbk6mKHBps27Zt9EfqSVvpcrsq3n333azlFc11+/bt\nTzjhhPCvhXX33HNPCOGQQw6JCvemWpyRal1pe+ebobbmOod2arGDtf6pqZYcvhNyGK5920cA\nACA3Arv9WGo1WYsWLcpt2rJly7Rp02rlKKWlpU899VTme1RDCG3atIne1Bmt/AppbyF8/PHH\nM+uPGDEiPz+/oKAgWh1TEytWrHj//fczy59//vkQQkFBweGHH155C9XqV0Wq1d/Ua1jnzJmT\ntbW5c+eGEJo0aRLdA5tyyCGHROfz6KOPhhCiO/5OP/30civLUu1XdEvgk08+OXfu3O3bt1fe\nqZo0mHqd63PPPZdZ/+9//3sVD53uvffeyzrX0Rs2CwoKMh+Tf9FFF4UQ3n333VdeeWX69Okh\nhPPOO6+6LzapuVqckWpdaXvnm6G25jqHdmqxg7X+qamWHL4TchiufdtHAAAgR0nShVCN/2os\nenNlCGHMmDEV1Yluz+zfv3/mphEjRkS7v/rqq+nlO3fuPPvss1OvEfjjH/9YxQZTW0844YTo\nn5s2bSoqKgohnHvuuZmVt2/fHr12sGfPnlFJWVlZ9LrDBg0avPfee+mVP/zww2jZ2lFHHVXF\nk4nOf9iwYemFqSeanX/++eXqv/baa9EtXSeffHJ6+amnnhpCKCgo2LVrV279qki1+puq3LBh\nw1WrVpVr6r777ov6dckll2Qe6Pbbbw8h1K1b9+WXX46qPfHEE5nVunbtGkLIy8t75plnym1K\nvWT29NNPTy+vfAqq2+CaNWuiXKx9+/Zbt25Nr79y5crUrXlDhw7NerhyKpnrV199NZrrk046\nKXPH7du3t2zZMoTQt2/fEEIikfjggw+qcsRI5gUTqdZnJ5LDjGRVrSttL3wzJHOd68wPdQ7t\n5NbBiqa1tuYoWf0xzOE7Ibdhr8U+AgAAe4fAbl+qYWD38MMPR7t379797bffTiaTO3bseOqp\np3r16lVUVPTII49EW0855ZSdO3dWpcFktp+Uv/jFL6J2Bg8e/PTTT69du3bnzp1ffPHFokWL\nTj755GjT1KlTU/WfeOKJqLBt27Z33333qlWr3n///XvuuSeKwBKJRHrSlHNgd+SRR9apU2fE\niBErVqxIJpM7d+58/PHHDz744GiXhQsXpu8ycuTIqPzaa6/96KOP3njjjS+//LK6/apItfo7\nd+7cKGb69re//dhjj23fvj2ZTK5du3bChAnFxcVRI+vXr888yvr166P3bBx//PEhhNatW5cL\nHSLPPfdcND4lJSU333zz2rVry8rKPv3007vuuiu6bzQ/P3/x4sXpu1Q+BTk0mHqyXq9evRYu\nXLhhw4bVq1dPnz794IMP/u53vxu98rWKgV0UTHTr1i0/Pz99rufMmVPRXKdcdtll4V9OPPHE\nqhwuJesFs9uxyhrY5TCAFan6lbZ3vhmSOc111g91ddvJrYMVTWstzlEOY5jDd0IOw16LfQQA\nAPYOgd2+VMPArqysLFozEmnYsGH0k6yoqGjmzJllZWWpWwWLi4uHDBmy2waT2X5Sbtu2bcCA\nAaFio0ePLtfIlClTCgoKMmsWFBTccccdVexdstLA7sc//vEf/vCH6IduUVFR+otEr7766nLt\nzJ8/v9yZrFq1Kod+VaTq/U0mk9OmTYsW94UQ8vLyot/kkfbt2y9btqyiowwaNChV87LLLquo\n2kMPPVS/fv1UzfT7QEtKSjLX5VU+BTk0uH79+mg5TzmtWrVaunRp9ASuM888s8LRTBPtOGrU\nqIkTJ2ad69///vcV7fvuu++mnqB/3333VeVwKVkvmGROcUyy+gNYiSpeaXvnmyGZ01xHFcp9\nqKvbTm4drGhaa3GOcrtCqvudkNtHrBavQwAAYC8Q2O1LNQzsksnk9u3br7vuuiOPPLK4uLh+\n/fqdOnX65S9/uXz58mjrP//5zx49ehQXF7ds2fKqq66qSoNZf1KWlZU99NBDQ4cO7dChQ716\n9fLy8ho2bPid73xnxIgRL7/8ctZ23n777VGjRnXq1Kl+/fqFhYXt2rW78MIL33zzzWr1Lutv\n++iX+VlnnZVMJp9++ukzzjijTZs2hYWFLVq0OOWUU+bNm5e1qSlTpnTq1KmwsLBJkyY9evQo\nLS3NrV8VqWJ/Ix988MGYMWO6du1aUlJSUFDQsmXLE0888dZbb43W+1QkeuNEZOnSpZXUXLNm\nzbhx47p37960adOCgoLWrVsfffTR48eP/+STTzIr7zawq26DyWRy69at48eP79atW4MGDerV\nq9exY8fRo0d//PHHyX/dmpf1PtbMRqLOjh07NplMPvXUU6effnrr1q0LCgoqn+uU6LlvTZs2\nrXxgs8p6weQWxySrP4CVqOKVtne+GZLVn+usH+oc2smhg8kKpjVSK3OU8xVS3e+E3D5itXgd\nAgAAe1oi+a9fULBfyM/P37Vr19ChQ1MPeIJMW7ZsOeSQQ0pLS0ePHj1hwoR9fToAAABQDd4S\nC3wNTZ48ubS0tE6dOqmHFQIAAMD+QmAHfN0sXbr0d7/7XQjhxz/+cbt27fb16QAAAED1COzY\nsxKJRCKRGD58eEzaYY/at9P02Wefffrpp3ffffcJJ5ywefPmpk2bXnPNNfvkTAAAAKAm8vf1\nCQDUgp07d7Zs2TL1z/r168+cObNNmzb78JQAAAAgNwK7vWrs2LHXXXddRVsLCgoaNWrUoUOH\nXr16nXfeeZ07d96b57aHXHnllSGE6MWFtdJOrN41cdNNN+3YseOyyy7b1ycSF7U13aH6Y5tI\nJFq0aLF+/fomTZr07dv3qquu6tKlS81PAwAAAPY+b4ndqyoP7NLVqVNnzJgx1157bSKR2NNn\nRW42btzYtGnTJk2arFu3bl+fy9eNsQUAAOCbzAq7f1OtcKwmUefNN9/cvXv3coVffvnlRx99\nNG/evHvvvXfHjh3XX399Xl6eh3DF1ksvvVRWVravz+LrydgCAADwTSaw2zc6d+7co0ePrJuG\nDx/+q1/96vjjj9+0adMNN9zwy1/+8oADDtjLp0dVvPDCC/v6FL62jC0AAADfZN4SG0ff/e53\nR44cGULYtWvX/PnzMyusXbt23Lhx3bt3b9q0aWFhYatWrfr06XPDDTds2rQps3LdunUTicS4\nceNCCAsXLhw4cGCrVq2Ki4s7dOgwZsyY0tLSqNo//vGPoUOHtm3btrCwsHXr1sOHD3/33Xez\nnt7cuXOHDx/esWPH+vXrFxQUNG/evHfv3hMmTNi6dWtm5ayvDU0/pbfeeuuCCy449NBD69at\nW1JS0qNHj0mTJmWurqqtdkIIO3bsuOWWW3r27NmkSZP69esffvjhv/3tbzds2BBCOOKIIxKJ\nxMknn5y14ymXXnppIpG46qqrQgilpaXRuXXo0CHamp+fH20tKysbP358+/bt8/Pzo8q5jWFu\n3Vy6dOmoUaO6du1aUlKSn5/fsGHDI444YuTIkW+88UbWTq1YseLiiy/u0qVLw4YN69at26FD\nhxEjRixbtqxctcp7lzlN27ZtiwqvuOKKEMKCBQvOOOOMtm3bFhUVtWjR4qSTTpo9e3bVxzaH\nfgEAAMD+J0maEKrxXw4uv/zyaNjnzZtXec2pU6dGNSdOnFhu0+zZsxs0aPD/2bu3GC3qw//j\nsyzLsspBWSz1QKlooWoVE0vRiEpTEQxIscaulgsTK8UI1ppGbcWmmqhJa7VR6ynURCttSUWs\nNvWABzxBql54QE0pWsUDYgXcBWSLuLv/i8n/+T3dXZeDoB/i6xUvJjPzzMx3Fm/emZlvubV3\n79677bZb5a+59957P/PMM532HzhwYFEU559//rx583r37vxM5eGHH97atgQUwwAAIABJREFU\n2nrnnXf26dOn06bGxsa33367+lBtbW2dklm1ESNGrFy5ssstLYqimDZtWreX9MADD+y+++5d\nD3X66afvpONs2LBh7NixlR369OlTfiXwa1/72uuvvz506NCiKE455ZSe/zqXXXbZkCFDyjtW\nU1MzZMiQIUOGHHnkkeXW8kouvPDCK664onKiX/7yl9t9D7djmDfddFNtbW1lh169/i/N19bW\n3nDDDZ32v/XWW+vq6roeua6u7pZbbqnes+fRdftnKleec845N998c/WVVJx//vlbeW+3dVwA\nAACwKxLs/kdOsJs9e3a557x586rXL168uAwWhx9++MMPP7x58+aOjo5Vq1b97ne/23PPPYui\nGDx48Lvvvlv9k0GDBhVFMWXKlAEDBpx77rmvvvrqxo0bn3/++WOPPbY8xaWXXtq/f/+mpqal\nS5du3Lhx+fLlp59+ernpvPPOqz7U1VdfXa4fN27cU0891dLSsmnTpmXLlpXPAxZFMWHChC63\ntJuCU17Sd7/73UGDBk2cOHHJkiXr169fs2bNvffeu++++5Y/efLJJ3fGcc4999xy/fjx4196\n6aWOjo7169fPmTNnwIABxxxzzODBg4uiaGpq6vmvU5owYUJRFI2NjZ3Wl33thz/84R577HHQ\nQQddeeWVN9544+OPP77d93Bbh7l8+fKyzI4aNWrhwoXNzc0dHR0ffPDBPffcc+ihh5ada+nS\npZX9H3zwwfIghx9++N/+9rdVq1atWLFi3rx5w4YNK6vZww8/vJWj6/bPVGa1o446qr6+/qST\nTlq8ePEHH3ywZs2aP/3pT0OGDCl/cu+9927x3m7ruAAAAGAXJdj9j5Bgt2LFirIc1dfXv/fe\ne9WbyjAxcuTIdevWdfrVs88+W5aRmTNnVq9vbGwsTzpr1qzq9W+++WZlCtrJkye3t7dXNv33\nv/8tf3XIIYdU/+SAAw4oiqJfv37r16/vdPapU6eWh3rzzTer13dbcCqXNGXKlLa2tupNd911\nV7npZz/72Q4/ztq1a8vnyEaOHNna2lq9/+OPP155XOtTBruyrzU2No4ePXrjxo2dtm7HPdzW\nYV5zzTXlyhdffLHTKdauXTtgwICBAwf+9re/Lde0t7ePGDGiKIqhQ4e2tLRU77xs2bL6+vqi\nKCoPuG1xdN3+mSrPxE2cOLHT9T/99NPlprFjx1av7/bebtO4AAAAYNflG3ZB1q1b99JLL11x\nxRVHHHHE6tWri6K44IILvvSlL1V2WLJkydKlS4uiuPjii/v379/p59/85jfLj6/NnTu360fN\namtry4+gVQwdOnT48OHl8iWXXFJTNUVufX19OSfG66+/Xv2ThQsX/uMf/1i4cGHlndyKiRMn\nlgsvvfTSVo63pqbm6quv7vSO5AknnFAuLF++fIcf54EHHti8eXNRFOecc07fvn2r9z/22GNP\nO+20rTzjFq+nKIo1a9b8+te/bmho6LR1u+/h1g9z3bp15UIZ16rtueeeq1atam5u/slPflKu\neeqpp/71r38VRXHOOecMGDCgeucRI0aceeaZ48aN22+//TZt2rQ1o+vZ5Zdf3un6v/Wtbx19\n9NFFUSxevHjt2rU9/3ybxgUAAAC7LsHu8zF+/PiaLgYOHHjooYdecsklZa2bNWvWZZddVv2r\nxx57rFyo/gpbtfIt15aWln/+85+dNh188MHV7a9UrmloaBg9enS3mzZu3Fjd/oYPHz5mzJij\njjqq66krDaVSVbbowAMPrJ5MoNSvX7/yK2ndTqDxKY/z/PPPlwuV14Gr7ahgVxo4cGC3Z9nu\ne7j1wzzssMPKhRkzZrz77rudftKpsj355JPlwnHHHdf1pDfeeOOiRYvuvPPO8lG7ik8aXQ8G\nDx58xBFHdF1fBruOjo4tpt5tGhcAAADsujpPQcDnbp999jnxxBOnT58+ZsyYTpuWLVtWLpSv\nVfZgxYoVBx98cPWavffeu+tu5af999prr65TAVTmoGhvb6/eunnz5vvvv/+RRx5ZtmxZc3Pz\npk2bOjo6iqJoaWkpd+j4/+9FblE5w0NXffv2/fDDD9va2nb4cd54441yofJoYbVuI9p2GzFi\nRLcTLBTbew+3fphTpkwZN27cY4899ve//33YsGHHHHPM+PHjx40bN3r06OoZG0rl43VFUey/\n//47ZHSfpNM/yIqvfvWr5cI777zT8xG2aVwAAACw6xLsPh8333xzpx63evXqSZMmffTRR0OG\nDLn55pu7TudaFEVzc/NWHr+Sfiq6nQN0i5s6Wbp06amnnlrphp/S1p93Rx1nw4YN5ULXF4qL\nomhsbOzVq1fXt4m3Tzk5Q1fbfQ+3fpi1tbX33XffJZdcctNNN7W2tj766KOPPvpoURR77LHH\n5MmTZ86cWb7vXKr8o+r6im4PPml0Pdhjjz26XV/5W2zxmcptGhcAAADsurwS+/k44IADDv9f\nxx9//AUXXFAUxXPPPVf5uH4nlWea3nvvvc09ampq2uHX3NLScsIJJ5Sl6cwzz3ziiSdWr15d\nTlPb0dFx55137vAz7nBljCtfQO52h219aqwH3T7z9Zndw4aGhquvvnrVqlVz586dNm1a+Xxl\nc3Pz3LlzjzrqqG4/9Lb1j0YWnzC6nlWe2eykUki35uZvx7gAAABglyPYBZk9e3b5euCll176\n6quvdt2h8oxSS0tL7x59UpD6NG677bZVq1YVRXHRRRfdeuutxxxzTGNjY+VJwNbW1h1+xh1u\nt912K4qio6Nj48aNXbc2Nzd//PHHO/UCPuN7OGDAgGnTps2dO3flypWvvPLK5ZdfXk44e+21\n195xxx3lPpV/VFv//Ob2+aSPG/b82GO3tmZcAAAAsOsS7II0NDRcf/31RVG0trb+6Ec/6rpD\n5StgL7zwwmd6ZUVRFMXTTz9dLvz4xz/uuvXll1/+bC9ne+yzzz7lwptvvtl167PPPruzL+Bz\nvIcHHXTQ7NmzX3jhhXJqi+uuu65cP3LkyHKh60QlO9Zrr73W7foVK1aUC9v0Eb2KTxoXAAAA\n7LoEuyyTJ0+eMmVKURSLFi36/e9/32lrZV7OT3p38uGHH164cOFHH320M66t8iTUXnvt1WnT\nhx9+ePvtt++Mk+5YhxxySLlQCWfV/vKXv+zsC/hs7uGaNWseeeSRrvOoFkWx7777lrOyrly5\nslxTmXH4/vvv77r/jBkzevfuXVdXVz4Y+Gm89tprr7/+etf1S5YsKYqirq6u8tf5JNs0LgAA\nANh1CXZxrrvuuvLNzQsuuKBTJRkzZsxhhx1WFMX8+fOfeOKJTj989913zzjjjAkTJpx22mk7\n48Iqj6ctXbq0en1bW9v06dMrr3NuceqAz9GECRPKhTlz5nT6ZNvTTz992223bdPRyo+yrVu3\nbuvnqfgM7uGGDRv23Xff448//uc//3nXrZs3b37llVeKohg2bFi5ZuzYsSNGjCiKYs6cOZ2C\n2ltvvTV//vy2trZRo0Z9+ctf3u5Lqrj88ss7rXnuuecWL15cFMX48eN33333yvqu93ZbxwUA\nAAC7LsHuf3R0bMN/O8mwYcNmz55dFEVzc/OsWbM6bb3xxhtra2vb29tPOumk66+//v333+/o\n6PjPf/5zxx13HH300StXruzdu/ell166My5s0qRJ5cLZZ5+9fPnyoig+/vjjRx999Nvf/vaC\nBQvmzp1bbr3nnnva2tp2xgV8evvvv//kyZOLoli8ePEZZ5xRvhjb2tr6hz/84cQTTzz55JOr\nm9EWDR06tCiKzZs3X3XVVe+8887SpUs3bdrU808+g3vYr1+/6dOnF0Vx++23n3rqqYsWLXr/\n/ffb2trWr1+/ZMmSqVOnlq+mnnXWWeX+NTU15YvYGzZsOPbYY+fOnfv222+/8cYbf/7zn7/9\n7W+vXbu2pqbmyiuv3L6LqTZq1Kjbb7/97LPP/ve//10URVtb2wMPPDB16tSynHbKcF3vbV1d\n3TaNCwAAAHZhHXyGLrroovK2P/TQQz3stmnTpq9//evlnnfffXenrQsWLKjuStVzaw4YMODB\nBx/stH/5Pf5JkyZ1PdFxxx1XFMUBBxzQddOMGTPKY1bmMC0rYeVc/fv3L6cKra+vnz9/fnt7\ne+UTew0NDaeeemr5q3LNtGnTul7ShAkTuh1+ufU73/lO9coddZy33nrrK1/5SmUU/fr1K2/g\nmDFjVq9eXd7Ypqambg/YyaJFizr93/TWW2/1fFXbdw+3dZitra0TJ07s4f/6888/v9NB5syZ\nU1dX13XPurq6W265pevpPuliuv0zlWP8wQ9+8Jvf/KacDqW+vr56ntkrrrhia+7tdowLAAAA\ndkWesEvUp0+fG264oVyeOXNmS0tL9daTTz75tdde+8UvfjF69OhBgwbV1tbus88+Y8aMufLK\nK5ctW3bCCSfspKuqqam56667fvWrX40aNaqhoaG9vf3AAw8877zzXnzxxVNOOaWmpuaPf/zj\nkUce2dDQ0L9//y1+j+zzst9++z333HMXXnjhyJEjGxoaevfufcQRR1x77bWPP/54Y2Nj+QJm\nZdrWno0bN27OnDkjR47s06fPnnvueeSRR5bvMvfgs7mHffv2ve+++xYsWNDU1HTggQfutttu\nvXr16t+//ze+8Y0ZM2Y888wz11xzTaefnHXWWS+//PLMmTNHjhy5++679+nTZ/jw4dOnT3/+\n+ee7nf9kO7S3t//0pz995JFHvve97w0ePLi2tnavvfaaNGnSQw89dPHFF3faudt7ux3jAgAA\ngF1RTcfOe7cTdimbNm3q27dvURSzZs0q3xJlh+jdu3dbW1tTU9O8efM+72sBAACAXYAn7PjC\n+eCDD95+++2u65999tly4aCDDvpsrwgAAADg/wh2fIGUr6AOGjRo7NixlSlZK6666qqiKHr1\n6tXzh9IAAAAAdirBji+QXr16NTU1FUWxYsWK8ePHP/bYYxs2bPj4449feOGF73//+/fee29R\nFGecccbw4cM/7ysFAAAAvrh8w44vls2bNzc1Nd19993dbp0wYcJdd91VPQkvn55v2AEAAMA2\n8YQdXyx1dXULFiz461//evLJJw8dOrRPnz719fX77bff1KlT58+ff99996l1AAAAwOfLE3YA\nAAAAEMQTdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAA\nAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAI\nItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7\nAAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAA\nAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAg\niGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHs\nAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAA\nAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACA\nIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESw\nAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAA\nAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAA\nggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDB\nDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEA\nAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAA\nCCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEE\nOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcA\nAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAA\nIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR\n7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0A\nAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAA\ngCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBE\nsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYA\nAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAA\nAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQ\nwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgB\nAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAA\nAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABB\nBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAH\nAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAA\nACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAE\nEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIId\nAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAA\nAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQ\nRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2\nAAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAA\nAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABA\nEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLY\nAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAA\nAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAA\nQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhg\nBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAA\nAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAA\nBBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCC\nHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMA\nAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAA\nEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIII\ndgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4A\nAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAA\nQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi\n2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsA\nAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAA\nAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCI\nYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewA\nAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAA\nAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAg\ngh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLAD\nAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAA\nABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACC\nCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEO\nAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAA\nAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAI\nItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7\nAAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAA\nAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAg\niGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHs\nAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAA\nAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACA\nIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESw\nAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAA\nAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAA\nggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDB\nDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEA\nAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAA\nCCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEE\nOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcA\nAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAA\nIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR\n7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0A\nAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAA\ngCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBE\nsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItjnHpZxAAAYeklEQVQBAAAA\nQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi\n2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsA\nAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAA\nAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAggh0AAAAABBHsAAAAACCI\nYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLADAAAAgCCCHQAAAAAEEewA\nAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAAABBEsAMAAACAIIIdAAAA\nAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACCCHYAAAAAEESwAwAAAIAg\ngh0AAAAABBHsAAAAACCIYAcAAAAAQQQ7AAAAAAgi2AEAAABAEMEOAAAAAIIIdgAAAAAQRLAD\nAAAAgCCCHQAAAAAEEewAAAAAIIhgBwAAAABBBDsAAAAACCLYAQAAAEAQwQ4AAAAAggh2AAAA\nABBEsAMAAACAIIIdAAAAAAQR7AAAAAAgiGAHAAAAAEEEOwAAAAAIItgBAAAAQBDBDgAAAACC\nCHYAAAAAEESwAwAAAIAggh0AAP+vHTsWAAAAABjkbz2NHYURAAAjwg4AAAAARoQdAAAAAIwI\nOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAA\nAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQd\nAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAA\nRoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4A\nAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAj\nwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAA\nAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFh\nBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAA\ngBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLAD\nAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADA\niLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEA\nAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE\n2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAA\nAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLs\nAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAA\nMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYA\nAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAY\nEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAA\nAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwI\nOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAA\nAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQd\nAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAA\nRoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4A\nAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAj\nwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAA\nAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFh\nBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAA\ngBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLAD\nAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADA\niLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEA\nAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE\n2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAA\nAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLs\nAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAA\nMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYA\nAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAY\nEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAA\nAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwI\nOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAA\nAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQd\nAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAA\nRoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4A\nAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAj\nwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAA\nAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFh\nBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAA\ngBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLAD\nAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADA\niLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEA\nAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE\n2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAA\nAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLs\nAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAA\nMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYA\nAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAY\nEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAA\nAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwI\nOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAA\nAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQd\nAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAA\nRoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4A\nAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAj\nwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAA\nAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFh\nBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAA\ngBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLAD\nAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADA\niLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEA\nAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE\n2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAA\nAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLs\nAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAA\nMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYA\nAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAY\nEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAA\nAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwI\nOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAA\nAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQd\nAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAA\nRoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4A\nAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAj\nwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAA\nAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFh\nBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAA\ngBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLAD\nAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADA\niLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEA\nAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE\n2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAA\nAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLs\nAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAA\nMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYA\nAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAY\nEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAA\nAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwI\nOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAA\nAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAARoQd\nAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4AAAAA\nRoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAjwg4A\nAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAAAAAj\nwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFhBwAA\nAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAAgBFh\nBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLADAAAA\ngBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGBE2AEAAADAiLAD\nAAAAgBFhBwAAAAAjwg4AAAAARoQdAAAAAIwIOwAAAAAYEXYAAAAAMCLsAAAAAGAk3idX+NwL\nPuQAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 924,
       "width": 840
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "i <- 1\n",
    "j <- 283\n",
    "tt <- extPreInf(tseCons, yTxps, trenDiNodes, trenDiDegNodes, txpMouse, genesMouse, i, j)\n",
    "pI <- plotIReps(yAll, tt$txpMin, tt$iNode, lp=\"left\")\n",
    "pTree <- plotTree(tt$tSub, 9, tt$minTInd, of=45, xlim=100)\n",
    "pgs <- parF(tt$g, tt$txShow, tt$tSub)\n",
    "\n",
    "options(repr.plot.width=14, repr.plot.height=15.4)\n",
    "library(plotgardener)\n",
    "pageCreate(width = 14, height = 12, showGuides = FALSE, default.units = \"inches\")\n",
    "par <- pgs[[1]]\n",
    "\n",
    "parText <- pgParams(\n",
    "    just = c(\"left\", \"top\"), fontsize = 18\n",
    ")#, fontface = \"bold\")\n",
    "\n",
    "hilite <- pgs[[2]]\n",
    "plotLegend(\n",
    "    legend = c(\"Transcripts covered by the candidate node\", \"Remaining transcripts\"),\n",
    "    fill = c(\"red\", \"blue\"), border = FALSE,\n",
    "    x = 0.5, y = 1, width = 1, height = 0.5,\n",
    "    just = c(\"left\", \"top\"), fontsize=18,\n",
    ")\n",
    "plotText(label=\"A) Transcripts mapping to gene SYK\", x = 0.5, y = 0.2,\n",
    "        params = parText)\n",
    "\n",
    "plotTranscripts(\n",
    "    params = par, x = 1.5, y = 5.5, width = 5, height = 8.5, \n",
    "    spaceHeight=2.5, boxHeight=0.15,\n",
    "    labels = \"transcript\", fill=\"blue\", \n",
    "    transcriptHighlights = hilite\n",
    ")\n",
    "\n",
    "plotGenomeLabel(\n",
    "  params = par, x = 0.5, y = 6.2, length=6,\n",
    "  just = c(\"left\", \"top\"), fontsize=16\n",
    ")\n",
    "\n",
    "plotText(label=\"B) Tree representing the transcripts covered by gene SYK\", \n",
    "         x = 7.1, y = 0.2, params = parText)\n",
    "plotGG(\n",
    "    plot = pTree,\n",
    "    x = 6.8, y =0.5, width = 8, height=6\n",
    ")\n",
    "\n",
    "plotText(label=\"C) Scaled inferential replicates for the transcript \\nENSMUST00000055087.6\", \n",
    "         x = 0.5, y = 7, params = parText)\n",
    "\n",
    "plotGG(\n",
    "    plot = pI[[1]],\n",
    "    x = 0.5, y = 7.6, width = 6, height=6\n",
    ")\n",
    "\n",
    "plotText(label=\"D) Scaled inferential replicates for the candidate node\", \n",
    "         x = 7.1, y = 7, params = parText)\n",
    "\n",
    "plotGG(\n",
    "    plot = pI[[2]],\n",
    "    x = 7.4, y = 7.6, width = 6, height=6\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "23e0d952",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages(source(\"../tree_helper.R\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d9ad804a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (chrom, chromstart = NULL, chromend = NULL, assembly = \"hg38\", \n",
       "<span style=white-space:pre-wrap>    fill = c(\"#669fd9\", \"#abcc8e\"), colorbyStrand = TRUE, strandSplit = FALSE, </span>\n",
       "<span style=white-space:pre-wrap>    boxHeight = unit(2, \"mm\"), spaceWidth = 0.02, spaceHeight = 0.3, </span>\n",
       "<span style=white-space:pre-wrap>    limitLabel = TRUE, fontsize = 8, labels = \"transcript\", stroke = 0.1, </span>\n",
       "<span style=white-space:pre-wrap>    bg = NA, x = NULL, y = NULL, width = NULL, height = NULL, </span>\n",
       "<span style=white-space:pre-wrap>    just = c(\"left\", \"top\"), default.units = \"inches\", draw = TRUE, </span>\n",
       "<span style=white-space:pre-wrap>    params = NULL) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    errorcheck_plotTranscripts &lt;- function(transcriptPlot, labels, </span>\n",
       "<span style=white-space:pre-wrap>        fill) {</span>\n",
       "<span style=white-space:pre-wrap>        regionErrors(chromstart = transcriptPlot$chromstart, </span>\n",
       "<span style=white-space:pre-wrap>            chromend = transcriptPlot$chromend)</span>\n",
       "<span style=white-space:pre-wrap>        if (!labels %in% c(NULL, \"transcript\", \"gene\", \"both\")) {</span>\n",
       "<span style=white-space:pre-wrap>            stop(\"Invalid 'labels' input. Options are 'NULL', \", </span>\n",
       "<span style=white-space:pre-wrap>                \"'transcript', 'gene', or 'both'.\", call. = FALSE)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        checkColorby(fill = fill, colorby = FALSE)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    strand_scale &lt;- function(strandSplit, height) {</span>\n",
       "<span style=white-space:pre-wrap>        if (strandSplit == TRUE) {</span>\n",
       "<span style=white-space:pre-wrap>            yscale &lt;- c(-height/2, height/2)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            yscale &lt;- c(0, height)</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        return(yscale)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    transcriptsInternal &lt;- parseParams(params = params, defaultArgs = formals(eval(match.call()[[1]])), </span>\n",
       "<span style=white-space:pre-wrap>        declaredArgs = lapply(match.call()[-1], eval.parent, </span>\n",
       "<span style=white-space:pre-wrap>            n = 2), class = \"transcriptsInternal\")</span>\n",
       "<span style=white-space:pre-wrap>    transcriptsInternal$just &lt;- justConversion(just = transcriptsInternal$just)</span>\n",
       "<span style=white-space:pre-wrap>    transcripts &lt;- structure(list(chrom = transcriptsInternal$chrom, </span>\n",
       "<span style=white-space:pre-wrap>        chromstart = transcriptsInternal$chromstart, chromend = transcriptsInternal$chromend, </span>\n",
       "<span style=white-space:pre-wrap>        assembly = transcriptsInternal$assembly, x = transcriptsInternal$x, </span>\n",
       "<span style=white-space:pre-wrap>        y = transcriptsInternal$y, width = transcriptsInternal$width, </span>\n",
       "<span style=white-space:pre-wrap>        height = transcriptsInternal$height, just = transcriptsInternal$just, </span>\n",
       "<span style=white-space:pre-wrap>        grobs = NULL), class = \"transcripts\")</span>\n",
       "<span style=white-space:pre-wrap>    attr(x = transcripts, which = \"plotted\") &lt;- transcriptsInternal$draw</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(transcripts$chrom)) {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"argument \\\"chrom\\\" is missing, with no default.\", </span>\n",
       "<span style=white-space:pre-wrap>            call. = FALSE)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    check_placement(object = transcripts)</span>\n",
       "<span style=white-space:pre-wrap>    errorcheck_plotTranscripts(transcriptPlot = transcripts, </span>\n",
       "<span style=white-space:pre-wrap>        labels = transcriptsInternal$labels, fill = transcriptsInternal$fill)</span>\n",
       "<span style=white-space:pre-wrap>    transcripts$assembly &lt;- parseAssembly(assembly = transcripts$assembly)</span>\n",
       "<span style=white-space:pre-wrap>    transcripts &lt;- defaultUnits(object = transcripts, default.units = transcriptsInternal$default.units)</span>\n",
       "<span style=white-space:pre-wrap>    transcriptsInternal$boxHeight &lt;- misc_defaultUnits(value = transcriptsInternal$boxHeight, </span>\n",
       "<span style=white-space:pre-wrap>        name = \"boxHeight\", default.units = transcriptsInternal$default.units)</span>\n",
       "<span style=white-space:pre-wrap>    buildData &lt;- geneData(object = transcripts, objectInternal = transcriptsInternal)</span>\n",
       "<span style=white-space:pre-wrap>    transcripts &lt;- buildData[[1]]</span>\n",
       "<span style=white-space:pre-wrap>    transcriptsInternal &lt;- buildData[[2]]</span>\n",
       "<span style=white-space:pre-wrap>    data &lt;- transcriptsInternal$data</span>\n",
       "<span style=white-space:pre-wrap>    data$length &lt;- data$TXEND - data$TXSTART</span>\n",
       "<span style=white-space:pre-wrap>    if (transcriptsInternal$colorbyStrand == TRUE) {</span>\n",
       "<span style=white-space:pre-wrap>        if (length(transcriptsInternal$fill) == 1) {</span>\n",
       "<span style=white-space:pre-wrap>            posCol &lt;- transcriptsInternal$fill</span>\n",
       "<span style=white-space:pre-wrap>            negCol &lt;- transcriptsInternal$fill</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            posCol &lt;- transcriptsInternal$fill[1]</span>\n",
       "<span style=white-space:pre-wrap>            negCol &lt;- transcriptsInternal$fill[2]</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        pos &lt;- data[which(data$TXSTRAND == \"+\"), ]</span>\n",
       "<span style=white-space:pre-wrap>        pos$color &lt;- rep(posCol, nrow(pos))</span>\n",
       "<span style=white-space:pre-wrap>        neg &lt;- data[which(data$TXSTRAND == \"-\"), ]</span>\n",
       "<span style=white-space:pre-wrap>        neg$color &lt;- rep(negCol, nrow(neg))</span>\n",
       "<span style=white-space:pre-wrap>        data &lt;- rbind(pos, neg)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        data$color &lt;- rep(transcriptsInternal$fill[1], nrow(data))</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (transcriptsInternal$strandSplit == TRUE) {</span>\n",
       "<span style=white-space:pre-wrap>        posStrand &lt;- data[which(data$TXSTRAND == \"+\"), ]</span>\n",
       "<span style=white-space:pre-wrap>        minStrand &lt;- data[which(data$TXSTRAND == \"-\"), ]</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    currentViewports &lt;- current_viewports()</span>\n",
       "<span style=white-space:pre-wrap>    vp_name &lt;- paste0(\"transcripts\", length(grep(pattern = \"transcripts\", </span>\n",
       "<span style=white-space:pre-wrap>        x = currentViewports)) + 1)</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(transcripts$x) | is.null(transcripts$y)) {</span>\n",
       "<span style=white-space:pre-wrap>        height &lt;- 0.5</span>\n",
       "<span style=white-space:pre-wrap>        yscale &lt;- strand_scale(strandSplit = transcriptsInternal$strandSplit, </span>\n",
       "<span style=white-space:pre-wrap>            height = height)</span>\n",
       "<span style=white-space:pre-wrap>        vp &lt;- viewport(height = unit(0.5, \"snpc\"), width = unit(1, </span>\n",
       "<span style=white-space:pre-wrap>            \"snpc\"), x = unit(0.5, \"npc\"), y = unit(0.5, \"npc\"), </span>\n",
       "<span style=white-space:pre-wrap>            clip = \"on\", xscale = transcriptsInternal$xscale, </span>\n",
       "<span style=white-space:pre-wrap>            yscale = yscale, just = \"center\", name = vp_name)</span>\n",
       "<span style=white-space:pre-wrap>        if (transcriptsInternal$draw == TRUE) {</span>\n",
       "<span style=white-space:pre-wrap>            vp$name &lt;- \"transcripts1\"</span>\n",
       "<span style=white-space:pre-wrap>            grid.newpage()</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        addViewport(vp_name)</span>\n",
       "<span style=white-space:pre-wrap>        page_coords &lt;- convert_page(object = transcripts)</span>\n",
       "<span style=white-space:pre-wrap>        height &lt;- convertHeight(page_coords$height, unitTo = get(\"page_units\", </span>\n",
       "<span style=white-space:pre-wrap>            envir = pgEnv), valueOnly = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>        yscale &lt;- strand_scale(strandSplit = transcriptsInternal$strandSplit, </span>\n",
       "<span style=white-space:pre-wrap>            height = height)</span>\n",
       "<span style=white-space:pre-wrap>        vp &lt;- viewport(height = page_coords$height, width = page_coords$width, </span>\n",
       "<span style=white-space:pre-wrap>            x = page_coords$x, y = page_coords$y, clip = \"on\", </span>\n",
       "<span style=white-space:pre-wrap>            xscale = transcriptsInternal$xscale, yscale = yscale, </span>\n",
       "<span style=white-space:pre-wrap>            just = transcriptsInternal$just, name = vp_name)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    backgroundGrob &lt;- rectGrob(gp = gpar(fill = transcriptsInternal$bg, </span>\n",
       "<span style=white-space:pre-wrap>        col = NA), name = \"background\")</span>\n",
       "<span style=white-space:pre-wrap>    assign(\"transcript_grobs\", gTree(vp = vp, children = gList(backgroundGrob)), </span>\n",
       "<span style=white-space:pre-wrap>        envir = pgEnv)</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(transcriptsInternal$labels)) {</span>\n",
       "<span style=white-space:pre-wrap>        textHeight &lt;- unit(0, \"npc\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        textHeight &lt;- heightDetails(textGrob(label = \"A\", gp = gpar(fontsize = transcriptsInternal$fontsize)))</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (is.null(transcripts$x) &amp; is.null(transcripts$y)) {</span>\n",
       "<span style=white-space:pre-wrap>        pushViewport(vp)</span>\n",
       "<span style=white-space:pre-wrap>        boxHeight &lt;- convertHeight(transcriptsInternal$boxHeight, </span>\n",
       "<span style=white-space:pre-wrap>            unitTo = \"npc\", valueOnly = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>        spaceHeight &lt;- boxHeight * (transcriptsInternal$spaceHeight)</span>\n",
       "<span style=white-space:pre-wrap>        textHeight &lt;- convertHeight(textHeight, unitTo = \"npc\", </span>\n",
       "<span style=white-space:pre-wrap>            valueOnly = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>        upViewport()</span>\n",
       "<span style=white-space:pre-wrap>        unit &lt;- \"npc\"</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        boxHeight &lt;- convertHeight(transcriptsInternal$boxHeight, </span>\n",
       "<span style=white-space:pre-wrap>            unitTo = get(\"page_units\", envir = pgEnv), valueOnly = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>        spaceHeight &lt;- boxHeight * (transcriptsInternal$spaceHeight)</span>\n",
       "<span style=white-space:pre-wrap>        textHeight &lt;- convertHeight(textHeight, unitTo = get(\"page_units\", </span>\n",
       "<span style=white-space:pre-wrap>            envir = pgEnv), valueOnly = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>        unit &lt;- get(\"page_units\", envir = pgEnv)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    maxRows &lt;- floor(height/(boxHeight + spaceHeight + textHeight + </span>\n",
       "<span style=white-space:pre-wrap>        0.25 * textHeight))</span>\n",
       "<span style=white-space:pre-wrap>    wiggle &lt;- abs(transcripts$chromend - transcripts$chromstart) * </span>\n",
       "<span style=white-space:pre-wrap>        transcriptsInternal$spaceWidth</span>\n",
       "<span style=white-space:pre-wrap>    if (transcriptsInternal$strandSplit == FALSE) {</span>\n",
       "<span style=white-space:pre-wrap>        repData &lt;- data[duplicated(data$TXNAME) == FALSE, ]</span>\n",
       "<span style=white-space:pre-wrap>        repData &lt;- defaultGenePriorities(data = repData, assembly = transcripts$assembly, </span>\n",
       "<span style=white-space:pre-wrap>            transcript = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>        rowData &lt;- assignRows(data = repData[, c(\"TXSTART\", \"TXEND\", </span>\n",
       "<span style=white-space:pre-wrap>            \"TXID\")], maxRows = maxRows, wiggle = wiggle, rowCol = 3, </span>\n",
       "<span style=white-space:pre-wrap>            limitLabel = transcriptsInternal$limitLabel, gTree = \"transcript_grobs\")</span>\n",
       "<span style=white-space:pre-wrap>        rowData &lt;- suppressMessages(dplyr::left_join(x = data, </span>\n",
       "<span style=white-space:pre-wrap>            y = rowData[, c(\"row\", \"TXID\")], by = \"TXID\"))</span>\n",
       "<span style=white-space:pre-wrap>        rowData$y &lt;- rowData$row * (boxHeight + spaceHeight + </span>\n",
       "<span style=white-space:pre-wrap>            textHeight + 0.25 * textHeight)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        repPos &lt;- posStrand[duplicated(posStrand$TXNAME) == FALSE, </span>\n",
       "<span style=white-space:pre-wrap>            ]</span>\n",
       "<span style=white-space:pre-wrap>        repMin &lt;- minStrand[duplicated(minStrand$TXNAME) == FALSE, </span>\n",
       "<span style=white-space:pre-wrap>            ]</span>\n",
       "<span style=white-space:pre-wrap>        repPos &lt;- defaultGenePriorities(data = repPos, assembly = transcripts$assembly, </span>\n",
       "<span style=white-space:pre-wrap>            transcript = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>        repMin &lt;- defaultGenePriorities(data = repMin, assembly = transcripts$assembly, </span>\n",
       "<span style=white-space:pre-wrap>            transcript = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>        posData &lt;- assignRows(data = repPos[, c(\"TXSTART\", \"TXEND\", </span>\n",
       "<span style=white-space:pre-wrap>            \"TXID\")], maxRows = floor(maxRows/2), wiggle = wiggle, </span>\n",
       "<span style=white-space:pre-wrap>            rowCol = 3, limitLabel = transcriptsInternal$limitLabel, </span>\n",
       "<span style=white-space:pre-wrap>            gTree = \"transcript_grobs\")</span>\n",
       "<span style=white-space:pre-wrap>        minData &lt;- assignRows(data = repMin[, c(\"TXSTART\", \"TXEND\", </span>\n",
       "<span style=white-space:pre-wrap>            \"TXID\")], maxRows = floor(maxRows/2), wiggle = wiggle, </span>\n",
       "<span style=white-space:pre-wrap>            rowCol = 3, limitLabel = transcriptsInternal$limitLabel, </span>\n",
       "<span style=white-space:pre-wrap>            gTree = \"transcript_grobs\", side = \"bottom\")</span>\n",
       "<span style=white-space:pre-wrap>        posData &lt;- suppressMessages(dplyr::left_join(x = posStrand, </span>\n",
       "<span style=white-space:pre-wrap>            y = posData[, c(\"row\", \"TXID\")], by = \"TXID\"))</span>\n",
       "<span style=white-space:pre-wrap>        minData &lt;- suppressMessages(dplyr::left_join(x = minStrand, </span>\n",
       "<span style=white-space:pre-wrap>            y = minData[, c(\"row\", \"TXID\")], by = \"TXID\"))</span>\n",
       "<span style=white-space:pre-wrap>        posData$y &lt;- (0.5 * spaceHeight) + posData$row * (boxHeight + </span>\n",
       "<span style=white-space:pre-wrap>            spaceHeight + textHeight + 0.25 * textHeight)</span>\n",
       "<span style=white-space:pre-wrap>        minData$y &lt;- ((0.5 * spaceHeight + boxHeight + textHeight + </span>\n",
       "<span style=white-space:pre-wrap>            0.25 * textHeight) + minData$row * (boxHeight + spaceHeight + </span>\n",
       "<span style=white-space:pre-wrap>            textHeight + 0.25 * textHeight)) * -1</span>\n",
       "<span style=white-space:pre-wrap>        rowData &lt;- rbind(posData, minData)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (nrow(rowData) &gt; 0) {</span>\n",
       "<span style=white-space:pre-wrap>        if ((transcripts$chromend - transcripts$chromstart) &gt;= </span>\n",
       "<span style=white-space:pre-wrap>            2.5e+07) {</span>\n",
       "<span style=white-space:pre-wrap>            transcriptLine &lt;- rectGrob(x = rowData$TXSTART, y = rowData$y, </span>\n",
       "<span style=white-space:pre-wrap>                width = rowData$TXEND - rowData$TXSTART, height = boxHeight, </span>\n",
       "<span style=white-space:pre-wrap>                just = c(\"left\", \"bottom\"), gp = gpar(fill = rowData$color, </span>\n",
       "<span style=white-space:pre-wrap>                  col = rowData$color, lwd = transcriptsInternal$stroke), </span>\n",
       "<span style=white-space:pre-wrap>                default.units = \"native\")</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        else {</span>\n",
       "<span style=white-space:pre-wrap>            transcriptLine &lt;- rectGrob(x = rowData$TXSTART, y = rowData$y + </span>\n",
       "<span style=white-space:pre-wrap>                0.5 * boxHeight, width = rowData$TXEND - rowData$TXSTART, </span>\n",
       "<span style=white-space:pre-wrap>                height = boxHeight * 0.2, just = \"left\", gp = gpar(fill = rowData$color, </span>\n",
       "<span style=white-space:pre-wrap>                  col = rowData$color, lwd = transcriptsInternal$stroke), </span>\n",
       "<span style=white-space:pre-wrap>                default.units = \"native\")</span>\n",
       "<span style=white-space:pre-wrap>            transcriptExons &lt;- rectGrob(x = rowData$EXONSTART, </span>\n",
       "<span style=white-space:pre-wrap>                y = rowData$y + 0.5 * boxHeight, width = rowData$EXONEND - </span>\n",
       "<span style=white-space:pre-wrap>                  rowData$EXONSTART, height = boxHeight * 0.65, </span>\n",
       "<span style=white-space:pre-wrap>                just = \"left\", gp = gpar(fill = rowData$color, </span>\n",
       "<span style=white-space:pre-wrap>                  col = NA), default.units = \"native\")</span>\n",
       "<span style=white-space:pre-wrap>            assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", </span>\n",
       "<span style=white-space:pre-wrap>                envir = pgEnv), child = transcriptExons), envir = pgEnv)</span>\n",
       "<span style=white-space:pre-wrap>            cdsData &lt;- rowData[which(!is.na(rowData$CDSSTART)), </span>\n",
       "<span style=white-space:pre-wrap>                ]</span>\n",
       "<span style=white-space:pre-wrap>            if (nrow(cdsData) &gt; 0) {</span>\n",
       "<span style=white-space:pre-wrap>                transcriptCds &lt;- rectGrob(x = cdsData$CDSSTART, </span>\n",
       "<span style=white-space:pre-wrap>                  y = cdsData$y + 0.5 * boxHeight, width = cdsData$CDSEND - </span>\n",
       "<span style=white-space:pre-wrap>                    cdsData$CDSSTART, height = boxHeight, just = \"left\", </span>\n",
       "<span style=white-space:pre-wrap>                  gp = gpar(fill = cdsData$color, col = cdsData$color, </span>\n",
       "<span style=white-space:pre-wrap>                    lwd = 1.25), default.units = \"native\")</span>\n",
       "<span style=white-space:pre-wrap>                assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", </span>\n",
       "<span style=white-space:pre-wrap>                  envir = pgEnv), child = transcriptCds), envir = pgEnv)</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>        assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", </span>\n",
       "<span style=white-space:pre-wrap>            envir = pgEnv), child = transcriptLine), envir = pgEnv)</span>\n",
       "<span style=white-space:pre-wrap>        if (!is.null(transcriptsInternal$labels)) {</span>\n",
       "<span style=white-space:pre-wrap>            transcriptNames &lt;- rowData[duplicated(rowData$TXNAME) == </span>\n",
       "<span style=white-space:pre-wrap>                FALSE, ]</span>\n",
       "<span style=white-space:pre-wrap>            transcriptNames$labelLoc &lt;- rowMeans(transcriptNames[c(\"TXSTART\", </span>\n",
       "<span style=white-space:pre-wrap>                \"TXEND\")])</span>\n",
       "<span style=white-space:pre-wrap>            if (transcriptsInternal$labels == \"transcript\") {</span>\n",
       "<span style=white-space:pre-wrap>                label &lt;- transcriptNames$TXNAME</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            else if (transcriptsInternal$labels == \"gene\") {</span>\n",
       "<span style=white-space:pre-wrap>                label &lt;- transcriptNames[[transcripts$assembly$display.column]]</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            else {</span>\n",
       "<span style=white-space:pre-wrap>                label &lt;- paste0(transcriptNames[[transcripts$assembly$display.column]], </span>\n",
       "<span style=white-space:pre-wrap>                  \":\", transcriptNames$TXNAME)</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>            checkedLabels &lt;- apply(data.frame(label = label, </span>\n",
       "<span style=white-space:pre-wrap>                labelLoc = transcriptNames$labelLoc), 1, cutoffLabel, </span>\n",
       "<span style=white-space:pre-wrap>                fontsize = transcriptsInternal$fontsize, xscale = transcriptsInternal$xscale, </span>\n",
       "<span style=white-space:pre-wrap>                vp = vp, unit = unit)</span>\n",
       "<span style=white-space:pre-wrap>            transcriptNames$label &lt;- checkedLabels</span>\n",
       "<span style=white-space:pre-wrap>            transcriptNames &lt;- transcriptNames[!is.na(transcriptNames$label), </span>\n",
       "<span style=white-space:pre-wrap>                ]</span>\n",
       "<span style=white-space:pre-wrap>            if (nrow(transcriptNames) &gt; 0) {</span>\n",
       "<span style=white-space:pre-wrap>                transcript_names &lt;- textGrob(label = transcriptNames$label, </span>\n",
       "<span style=white-space:pre-wrap>                  x = transcriptNames$labelLoc, y = transcriptNames$y + </span>\n",
       "<span style=white-space:pre-wrap>                    boxHeight + textHeight * 0.25, just = \"bottom\", </span>\n",
       "<span style=white-space:pre-wrap>                  gp = gpar(col = transcriptNames$color, fontsize = transcriptsInternal$fontsize), </span>\n",
       "<span style=white-space:pre-wrap>                  default.units = \"native\", check.overlap = TRUE)</span>\n",
       "<span style=white-space:pre-wrap>                assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", </span>\n",
       "<span style=white-space:pre-wrap>                  envir = pgEnv), child = transcript_names), </span>\n",
       "<span style=white-space:pre-wrap>                  envir = pgEnv)</span>\n",
       "<span style=white-space:pre-wrap>            }</span>\n",
       "<span style=white-space:pre-wrap>        }</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (transcriptsInternal$draw == TRUE) {</span>\n",
       "<span style=white-space:pre-wrap>        grid.draw(get(\"transcript_grobs\", envir = pgEnv))</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    transcripts$grobs &lt;- get(\"transcript_grobs\", envir = pgEnv)</span>\n",
       "<span style=white-space:pre-wrap>    message(\"transcripts[\", vp$name, \"]\")</span>\n",
       "<span style=white-space:pre-wrap>    invisible(transcripts)</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (chrom, chromstart = NULL, chromend = NULL, assembly = \"hg38\", \n",
       "    fill = c(\"\\#669fd9\", \"\\#abcc8e\"), colorbyStrand = TRUE, strandSplit = FALSE, \n",
       "    boxHeight = unit(2, \"mm\"), spaceWidth = 0.02, spaceHeight = 0.3, \n",
       "    limitLabel = TRUE, fontsize = 8, labels = \"transcript\", stroke = 0.1, \n",
       "    bg = NA, x = NULL, y = NULL, width = NULL, height = NULL, \n",
       "    just = c(\"left\", \"top\"), default.units = \"inches\", draw = TRUE, \n",
       "    params = NULL) \n",
       "\\{\n",
       "    errorcheck\\_plotTranscripts <- function(transcriptPlot, labels, \n",
       "        fill) \\{\n",
       "        regionErrors(chromstart = transcriptPlot\\$chromstart, \n",
       "            chromend = transcriptPlot\\$chromend)\n",
       "        if (!labels \\%in\\% c(NULL, \"transcript\", \"gene\", \"both\")) \\{\n",
       "            stop(\"Invalid 'labels' input. Options are 'NULL', \", \n",
       "                \"'transcript', 'gene', or 'both'.\", call. = FALSE)\n",
       "        \\}\n",
       "        checkColorby(fill = fill, colorby = FALSE)\n",
       "    \\}\n",
       "    strand\\_scale <- function(strandSplit, height) \\{\n",
       "        if (strandSplit == TRUE) \\{\n",
       "            yscale <- c(-height/2, height/2)\n",
       "        \\}\n",
       "        else \\{\n",
       "            yscale <- c(0, height)\n",
       "        \\}\n",
       "        return(yscale)\n",
       "    \\}\n",
       "    transcriptsInternal <- parseParams(params = params, defaultArgs = formals(eval(match.call(){[}{[}1{]}{]})), \n",
       "        declaredArgs = lapply(match.call(){[}-1{]}, eval.parent, \n",
       "            n = 2), class = \"transcriptsInternal\")\n",
       "    transcriptsInternal\\$just <- justConversion(just = transcriptsInternal\\$just)\n",
       "    transcripts <- structure(list(chrom = transcriptsInternal\\$chrom, \n",
       "        chromstart = transcriptsInternal\\$chromstart, chromend = transcriptsInternal\\$chromend, \n",
       "        assembly = transcriptsInternal\\$assembly, x = transcriptsInternal\\$x, \n",
       "        y = transcriptsInternal\\$y, width = transcriptsInternal\\$width, \n",
       "        height = transcriptsInternal\\$height, just = transcriptsInternal\\$just, \n",
       "        grobs = NULL), class = \"transcripts\")\n",
       "    attr(x = transcripts, which = \"plotted\") <- transcriptsInternal\\$draw\n",
       "    if (is.null(transcripts\\$chrom)) \\{\n",
       "        stop(\"argument \\textbackslash{}\"chrom\\textbackslash{}\" is missing, with no default.\", \n",
       "            call. = FALSE)\n",
       "    \\}\n",
       "    check\\_placement(object = transcripts)\n",
       "    errorcheck\\_plotTranscripts(transcriptPlot = transcripts, \n",
       "        labels = transcriptsInternal\\$labels, fill = transcriptsInternal\\$fill)\n",
       "    transcripts\\$assembly <- parseAssembly(assembly = transcripts\\$assembly)\n",
       "    transcripts <- defaultUnits(object = transcripts, default.units = transcriptsInternal\\$default.units)\n",
       "    transcriptsInternal\\$boxHeight <- misc\\_defaultUnits(value = transcriptsInternal\\$boxHeight, \n",
       "        name = \"boxHeight\", default.units = transcriptsInternal\\$default.units)\n",
       "    buildData <- geneData(object = transcripts, objectInternal = transcriptsInternal)\n",
       "    transcripts <- buildData{[}{[}1{]}{]}\n",
       "    transcriptsInternal <- buildData{[}{[}2{]}{]}\n",
       "    data <- transcriptsInternal\\$data\n",
       "    data\\$length <- data\\$TXEND - data\\$TXSTART\n",
       "    if (transcriptsInternal\\$colorbyStrand == TRUE) \\{\n",
       "        if (length(transcriptsInternal\\$fill) == 1) \\{\n",
       "            posCol <- transcriptsInternal\\$fill\n",
       "            negCol <- transcriptsInternal\\$fill\n",
       "        \\}\n",
       "        else \\{\n",
       "            posCol <- transcriptsInternal\\$fill{[}1{]}\n",
       "            negCol <- transcriptsInternal\\$fill{[}2{]}\n",
       "        \\}\n",
       "        pos <- data{[}which(data\\$TXSTRAND == \"+\"), {]}\n",
       "        pos\\$color <- rep(posCol, nrow(pos))\n",
       "        neg <- data{[}which(data\\$TXSTRAND == \"-\"), {]}\n",
       "        neg\\$color <- rep(negCol, nrow(neg))\n",
       "        data <- rbind(pos, neg)\n",
       "    \\}\n",
       "    else \\{\n",
       "        data\\$color <- rep(transcriptsInternal\\$fill{[}1{]}, nrow(data))\n",
       "    \\}\n",
       "    if (transcriptsInternal\\$strandSplit == TRUE) \\{\n",
       "        posStrand <- data{[}which(data\\$TXSTRAND == \"+\"), {]}\n",
       "        minStrand <- data{[}which(data\\$TXSTRAND == \"-\"), {]}\n",
       "    \\}\n",
       "    currentViewports <- current\\_viewports()\n",
       "    vp\\_name <- paste0(\"transcripts\", length(grep(pattern = \"transcripts\", \n",
       "        x = currentViewports)) + 1)\n",
       "    if (is.null(transcripts\\$x) \\textbar{} is.null(transcripts\\$y)) \\{\n",
       "        height <- 0.5\n",
       "        yscale <- strand\\_scale(strandSplit = transcriptsInternal\\$strandSplit, \n",
       "            height = height)\n",
       "        vp <- viewport(height = unit(0.5, \"snpc\"), width = unit(1, \n",
       "            \"snpc\"), x = unit(0.5, \"npc\"), y = unit(0.5, \"npc\"), \n",
       "            clip = \"on\", xscale = transcriptsInternal\\$xscale, \n",
       "            yscale = yscale, just = \"center\", name = vp\\_name)\n",
       "        if (transcriptsInternal\\$draw == TRUE) \\{\n",
       "            vp\\$name <- \"transcripts1\"\n",
       "            grid.newpage()\n",
       "        \\}\n",
       "    \\}\n",
       "    else \\{\n",
       "        addViewport(vp\\_name)\n",
       "        page\\_coords <- convert\\_page(object = transcripts)\n",
       "        height <- convertHeight(page\\_coords\\$height, unitTo = get(\"page\\_units\", \n",
       "            envir = pgEnv), valueOnly = TRUE)\n",
       "        yscale <- strand\\_scale(strandSplit = transcriptsInternal\\$strandSplit, \n",
       "            height = height)\n",
       "        vp <- viewport(height = page\\_coords\\$height, width = page\\_coords\\$width, \n",
       "            x = page\\_coords\\$x, y = page\\_coords\\$y, clip = \"on\", \n",
       "            xscale = transcriptsInternal\\$xscale, yscale = yscale, \n",
       "            just = transcriptsInternal\\$just, name = vp\\_name)\n",
       "    \\}\n",
       "    backgroundGrob <- rectGrob(gp = gpar(fill = transcriptsInternal\\$bg, \n",
       "        col = NA), name = \"background\")\n",
       "    assign(\"transcript\\_grobs\", gTree(vp = vp, children = gList(backgroundGrob)), \n",
       "        envir = pgEnv)\n",
       "    if (is.null(transcriptsInternal\\$labels)) \\{\n",
       "        textHeight <- unit(0, \"npc\")\n",
       "    \\}\n",
       "    else \\{\n",
       "        textHeight <- heightDetails(textGrob(label = \"A\", gp = gpar(fontsize = transcriptsInternal\\$fontsize)))\n",
       "    \\}\n",
       "    if (is.null(transcripts\\$x) \\& is.null(transcripts\\$y)) \\{\n",
       "        pushViewport(vp)\n",
       "        boxHeight <- convertHeight(transcriptsInternal\\$boxHeight, \n",
       "            unitTo = \"npc\", valueOnly = TRUE)\n",
       "        spaceHeight <- boxHeight * (transcriptsInternal\\$spaceHeight)\n",
       "        textHeight <- convertHeight(textHeight, unitTo = \"npc\", \n",
       "            valueOnly = TRUE)\n",
       "        upViewport()\n",
       "        unit <- \"npc\"\n",
       "    \\}\n",
       "    else \\{\n",
       "        boxHeight <- convertHeight(transcriptsInternal\\$boxHeight, \n",
       "            unitTo = get(\"page\\_units\", envir = pgEnv), valueOnly = TRUE)\n",
       "        spaceHeight <- boxHeight * (transcriptsInternal\\$spaceHeight)\n",
       "        textHeight <- convertHeight(textHeight, unitTo = get(\"page\\_units\", \n",
       "            envir = pgEnv), valueOnly = TRUE)\n",
       "        unit <- get(\"page\\_units\", envir = pgEnv)\n",
       "    \\}\n",
       "    maxRows <- floor(height/(boxHeight + spaceHeight + textHeight + \n",
       "        0.25 * textHeight))\n",
       "    wiggle <- abs(transcripts\\$chromend - transcripts\\$chromstart) * \n",
       "        transcriptsInternal\\$spaceWidth\n",
       "    if (transcriptsInternal\\$strandSplit == FALSE) \\{\n",
       "        repData <- data{[}duplicated(data\\$TXNAME) == FALSE, {]}\n",
       "        repData <- defaultGenePriorities(data = repData, assembly = transcripts\\$assembly, \n",
       "            transcript = TRUE)\n",
       "        rowData <- assignRows(data = repData{[}, c(\"TXSTART\", \"TXEND\", \n",
       "            \"TXID\"){]}, maxRows = maxRows, wiggle = wiggle, rowCol = 3, \n",
       "            limitLabel = transcriptsInternal\\$limitLabel, gTree = \"transcript\\_grobs\")\n",
       "        rowData <- suppressMessages(dplyr::left\\_join(x = data, \n",
       "            y = rowData{[}, c(\"row\", \"TXID\"){]}, by = \"TXID\"))\n",
       "        rowData\\$y <- rowData\\$row * (boxHeight + spaceHeight + \n",
       "            textHeight + 0.25 * textHeight)\n",
       "    \\}\n",
       "    else \\{\n",
       "        repPos <- posStrand{[}duplicated(posStrand\\$TXNAME) == FALSE, \n",
       "            {]}\n",
       "        repMin <- minStrand{[}duplicated(minStrand\\$TXNAME) == FALSE, \n",
       "            {]}\n",
       "        repPos <- defaultGenePriorities(data = repPos, assembly = transcripts\\$assembly, \n",
       "            transcript = TRUE)\n",
       "        repMin <- defaultGenePriorities(data = repMin, assembly = transcripts\\$assembly, \n",
       "            transcript = TRUE)\n",
       "        posData <- assignRows(data = repPos{[}, c(\"TXSTART\", \"TXEND\", \n",
       "            \"TXID\"){]}, maxRows = floor(maxRows/2), wiggle = wiggle, \n",
       "            rowCol = 3, limitLabel = transcriptsInternal\\$limitLabel, \n",
       "            gTree = \"transcript\\_grobs\")\n",
       "        minData <- assignRows(data = repMin{[}, c(\"TXSTART\", \"TXEND\", \n",
       "            \"TXID\"){]}, maxRows = floor(maxRows/2), wiggle = wiggle, \n",
       "            rowCol = 3, limitLabel = transcriptsInternal\\$limitLabel, \n",
       "            gTree = \"transcript\\_grobs\", side = \"bottom\")\n",
       "        posData <- suppressMessages(dplyr::left\\_join(x = posStrand, \n",
       "            y = posData{[}, c(\"row\", \"TXID\"){]}, by = \"TXID\"))\n",
       "        minData <- suppressMessages(dplyr::left\\_join(x = minStrand, \n",
       "            y = minData{[}, c(\"row\", \"TXID\"){]}, by = \"TXID\"))\n",
       "        posData\\$y <- (0.5 * spaceHeight) + posData\\$row * (boxHeight + \n",
       "            spaceHeight + textHeight + 0.25 * textHeight)\n",
       "        minData\\$y <- ((0.5 * spaceHeight + boxHeight + textHeight + \n",
       "            0.25 * textHeight) + minData\\$row * (boxHeight + spaceHeight + \n",
       "            textHeight + 0.25 * textHeight)) * -1\n",
       "        rowData <- rbind(posData, minData)\n",
       "    \\}\n",
       "    if (nrow(rowData) > 0) \\{\n",
       "        if ((transcripts\\$chromend - transcripts\\$chromstart) >= \n",
       "            2.5e+07) \\{\n",
       "            transcriptLine <- rectGrob(x = rowData\\$TXSTART, y = rowData\\$y, \n",
       "                width = rowData\\$TXEND - rowData\\$TXSTART, height = boxHeight, \n",
       "                just = c(\"left\", \"bottom\"), gp = gpar(fill = rowData\\$color, \n",
       "                  col = rowData\\$color, lwd = transcriptsInternal\\$stroke), \n",
       "                default.units = \"native\")\n",
       "        \\}\n",
       "        else \\{\n",
       "            transcriptLine <- rectGrob(x = rowData\\$TXSTART, y = rowData\\$y + \n",
       "                0.5 * boxHeight, width = rowData\\$TXEND - rowData\\$TXSTART, \n",
       "                height = boxHeight * 0.2, just = \"left\", gp = gpar(fill = rowData\\$color, \n",
       "                  col = rowData\\$color, lwd = transcriptsInternal\\$stroke), \n",
       "                default.units = \"native\")\n",
       "            transcriptExons <- rectGrob(x = rowData\\$EXONSTART, \n",
       "                y = rowData\\$y + 0.5 * boxHeight, width = rowData\\$EXONEND - \n",
       "                  rowData\\$EXONSTART, height = boxHeight * 0.65, \n",
       "                just = \"left\", gp = gpar(fill = rowData\\$color, \n",
       "                  col = NA), default.units = \"native\")\n",
       "            assign(\"transcript\\_grobs\", addGrob(get(\"transcript\\_grobs\", \n",
       "                envir = pgEnv), child = transcriptExons), envir = pgEnv)\n",
       "            cdsData <- rowData{[}which(!is.na(rowData\\$CDSSTART)), \n",
       "                {]}\n",
       "            if (nrow(cdsData) > 0) \\{\n",
       "                transcriptCds <- rectGrob(x = cdsData\\$CDSSTART, \n",
       "                  y = cdsData\\$y + 0.5 * boxHeight, width = cdsData\\$CDSEND - \n",
       "                    cdsData\\$CDSSTART, height = boxHeight, just = \"left\", \n",
       "                  gp = gpar(fill = cdsData\\$color, col = cdsData\\$color, \n",
       "                    lwd = 1.25), default.units = \"native\")\n",
       "                assign(\"transcript\\_grobs\", addGrob(get(\"transcript\\_grobs\", \n",
       "                  envir = pgEnv), child = transcriptCds), envir = pgEnv)\n",
       "            \\}\n",
       "        \\}\n",
       "        assign(\"transcript\\_grobs\", addGrob(get(\"transcript\\_grobs\", \n",
       "            envir = pgEnv), child = transcriptLine), envir = pgEnv)\n",
       "        if (!is.null(transcriptsInternal\\$labels)) \\{\n",
       "            transcriptNames <- rowData{[}duplicated(rowData\\$TXNAME) == \n",
       "                FALSE, {]}\n",
       "            transcriptNames\\$labelLoc <- rowMeans(transcriptNames{[}c(\"TXSTART\", \n",
       "                \"TXEND\"){]})\n",
       "            if (transcriptsInternal\\$labels == \"transcript\") \\{\n",
       "                label <- transcriptNames\\$TXNAME\n",
       "            \\}\n",
       "            else if (transcriptsInternal\\$labels == \"gene\") \\{\n",
       "                label <- transcriptNames{[}{[}transcripts\\$assembly\\$display.column{]}{]}\n",
       "            \\}\n",
       "            else \\{\n",
       "                label <- paste0(transcriptNames{[}{[}transcripts\\$assembly\\$display.column{]}{]}, \n",
       "                  \":\", transcriptNames\\$TXNAME)\n",
       "            \\}\n",
       "            checkedLabels <- apply(data.frame(label = label, \n",
       "                labelLoc = transcriptNames\\$labelLoc), 1, cutoffLabel, \n",
       "                fontsize = transcriptsInternal\\$fontsize, xscale = transcriptsInternal\\$xscale, \n",
       "                vp = vp, unit = unit)\n",
       "            transcriptNames\\$label <- checkedLabels\n",
       "            transcriptNames <- transcriptNames{[}!is.na(transcriptNames\\$label), \n",
       "                {]}\n",
       "            if (nrow(transcriptNames) > 0) \\{\n",
       "                transcript\\_names <- textGrob(label = transcriptNames\\$label, \n",
       "                  x = transcriptNames\\$labelLoc, y = transcriptNames\\$y + \n",
       "                    boxHeight + textHeight * 0.25, just = \"bottom\", \n",
       "                  gp = gpar(col = transcriptNames\\$color, fontsize = transcriptsInternal\\$fontsize), \n",
       "                  default.units = \"native\", check.overlap = TRUE)\n",
       "                assign(\"transcript\\_grobs\", addGrob(get(\"transcript\\_grobs\", \n",
       "                  envir = pgEnv), child = transcript\\_names), \n",
       "                  envir = pgEnv)\n",
       "            \\}\n",
       "        \\}\n",
       "    \\}\n",
       "    if (transcriptsInternal\\$draw == TRUE) \\{\n",
       "        grid.draw(get(\"transcript\\_grobs\", envir = pgEnv))\n",
       "    \\}\n",
       "    transcripts\\$grobs <- get(\"transcript\\_grobs\", envir = pgEnv)\n",
       "    message(\"transcripts{[}\", vp\\$name, \"{]}\")\n",
       "    invisible(transcripts)\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (chrom, chromstart = NULL, chromend = NULL, assembly = \"hg38\", \n",
       "    fill = c(\"#669fd9\", \"#abcc8e\"), colorbyStrand = TRUE, strandSplit = FALSE, \n",
       "    boxHeight = unit(2, \"mm\"), spaceWidth = 0.02, spaceHeight = 0.3, \n",
       "    limitLabel = TRUE, fontsize = 8, labels = \"transcript\", stroke = 0.1, \n",
       "    bg = NA, x = NULL, y = NULL, width = NULL, height = NULL, \n",
       "    just = c(\"left\", \"top\"), default.units = \"inches\", draw = TRUE, \n",
       "    params = NULL) \n",
       "{\n",
       "    errorcheck_plotTranscripts <- function(transcriptPlot, labels, \n",
       "        fill) {\n",
       "        regionErrors(chromstart = transcriptPlot$chromstart, \n",
       "            chromend = transcriptPlot$chromend)\n",
       "        if (!labels %in% c(NULL, \"transcript\", \"gene\", \"both\")) {\n",
       "            stop(\"Invalid 'labels' input. Options are 'NULL', \", \n",
       "                \"'transcript', 'gene', or 'both'.\", call. = FALSE)\n",
       "        }\n",
       "        checkColorby(fill = fill, colorby = FALSE)\n",
       "    }\n",
       "    strand_scale <- function(strandSplit, height) {\n",
       "        if (strandSplit == TRUE) {\n",
       "            yscale <- c(-height/2, height/2)\n",
       "        }\n",
       "        else {\n",
       "            yscale <- c(0, height)\n",
       "        }\n",
       "        return(yscale)\n",
       "    }\n",
       "    transcriptsInternal <- parseParams(params = params, defaultArgs = formals(eval(match.call()[[1]])), \n",
       "        declaredArgs = lapply(match.call()[-1], eval.parent, \n",
       "            n = 2), class = \"transcriptsInternal\")\n",
       "    transcriptsInternal$just <- justConversion(just = transcriptsInternal$just)\n",
       "    transcripts <- structure(list(chrom = transcriptsInternal$chrom, \n",
       "        chromstart = transcriptsInternal$chromstart, chromend = transcriptsInternal$chromend, \n",
       "        assembly = transcriptsInternal$assembly, x = transcriptsInternal$x, \n",
       "        y = transcriptsInternal$y, width = transcriptsInternal$width, \n",
       "        height = transcriptsInternal$height, just = transcriptsInternal$just, \n",
       "        grobs = NULL), class = \"transcripts\")\n",
       "    attr(x = transcripts, which = \"plotted\") <- transcriptsInternal$draw\n",
       "    if (is.null(transcripts$chrom)) {\n",
       "        stop(\"argument \\\"chrom\\\" is missing, with no default.\", \n",
       "            call. = FALSE)\n",
       "    }\n",
       "    check_placement(object = transcripts)\n",
       "    errorcheck_plotTranscripts(transcriptPlot = transcripts, \n",
       "        labels = transcriptsInternal$labels, fill = transcriptsInternal$fill)\n",
       "    transcripts$assembly <- parseAssembly(assembly = transcripts$assembly)\n",
       "    transcripts <- defaultUnits(object = transcripts, default.units = transcriptsInternal$default.units)\n",
       "    transcriptsInternal$boxHeight <- misc_defaultUnits(value = transcriptsInternal$boxHeight, \n",
       "        name = \"boxHeight\", default.units = transcriptsInternal$default.units)\n",
       "    buildData <- geneData(object = transcripts, objectInternal = transcriptsInternal)\n",
       "    transcripts <- buildData[[1]]\n",
       "    transcriptsInternal <- buildData[[2]]\n",
       "    data <- transcriptsInternal$data\n",
       "    data$length <- data$TXEND - data$TXSTART\n",
       "    if (transcriptsInternal$colorbyStrand == TRUE) {\n",
       "        if (length(transcriptsInternal$fill) == 1) {\n",
       "            posCol <- transcriptsInternal$fill\n",
       "            negCol <- transcriptsInternal$fill\n",
       "        }\n",
       "        else {\n",
       "            posCol <- transcriptsInternal$fill[1]\n",
       "            negCol <- transcriptsInternal$fill[2]\n",
       "        }\n",
       "        pos <- data[which(data$TXSTRAND == \"+\"), ]\n",
       "        pos$color <- rep(posCol, nrow(pos))\n",
       "        neg <- data[which(data$TXSTRAND == \"-\"), ]\n",
       "        neg$color <- rep(negCol, nrow(neg))\n",
       "        data <- rbind(pos, neg)\n",
       "    }\n",
       "    else {\n",
       "        data$color <- rep(transcriptsInternal$fill[1], nrow(data))\n",
       "    }\n",
       "    if (transcriptsInternal$strandSplit == TRUE) {\n",
       "        posStrand <- data[which(data$TXSTRAND == \"+\"), ]\n",
       "        minStrand <- data[which(data$TXSTRAND == \"-\"), ]\n",
       "    }\n",
       "    currentViewports <- current_viewports()\n",
       "    vp_name <- paste0(\"transcripts\", length(grep(pattern = \"transcripts\", \n",
       "        x = currentViewports)) + 1)\n",
       "    if (is.null(transcripts$x) | is.null(transcripts$y)) {\n",
       "        height <- 0.5\n",
       "        yscale <- strand_scale(strandSplit = transcriptsInternal$strandSplit, \n",
       "            height = height)\n",
       "        vp <- viewport(height = unit(0.5, \"snpc\"), width = unit(1, \n",
       "            \"snpc\"), x = unit(0.5, \"npc\"), y = unit(0.5, \"npc\"), \n",
       "            clip = \"on\", xscale = transcriptsInternal$xscale, \n",
       "            yscale = yscale, just = \"center\", name = vp_name)\n",
       "        if (transcriptsInternal$draw == TRUE) {\n",
       "            vp$name <- \"transcripts1\"\n",
       "            grid.newpage()\n",
       "        }\n",
       "    }\n",
       "    else {\n",
       "        addViewport(vp_name)\n",
       "        page_coords <- convert_page(object = transcripts)\n",
       "        height <- convertHeight(page_coords$height, unitTo = get(\"page_units\", \n",
       "            envir = pgEnv), valueOnly = TRUE)\n",
       "        yscale <- strand_scale(strandSplit = transcriptsInternal$strandSplit, \n",
       "            height = height)\n",
       "        vp <- viewport(height = page_coords$height, width = page_coords$width, \n",
       "            x = page_coords$x, y = page_coords$y, clip = \"on\", \n",
       "            xscale = transcriptsInternal$xscale, yscale = yscale, \n",
       "            just = transcriptsInternal$just, name = vp_name)\n",
       "    }\n",
       "    backgroundGrob <- rectGrob(gp = gpar(fill = transcriptsInternal$bg, \n",
       "        col = NA), name = \"background\")\n",
       "    assign(\"transcript_grobs\", gTree(vp = vp, children = gList(backgroundGrob)), \n",
       "        envir = pgEnv)\n",
       "    if (is.null(transcriptsInternal$labels)) {\n",
       "        textHeight <- unit(0, \"npc\")\n",
       "    }\n",
       "    else {\n",
       "        textHeight <- heightDetails(textGrob(label = \"A\", gp = gpar(fontsize = transcriptsInternal$fontsize)))\n",
       "    }\n",
       "    if (is.null(transcripts$x) & is.null(transcripts$y)) {\n",
       "        pushViewport(vp)\n",
       "        boxHeight <- convertHeight(transcriptsInternal$boxHeight, \n",
       "            unitTo = \"npc\", valueOnly = TRUE)\n",
       "        spaceHeight <- boxHeight * (transcriptsInternal$spaceHeight)\n",
       "        textHeight <- convertHeight(textHeight, unitTo = \"npc\", \n",
       "            valueOnly = TRUE)\n",
       "        upViewport()\n",
       "        unit <- \"npc\"\n",
       "    }\n",
       "    else {\n",
       "        boxHeight <- convertHeight(transcriptsInternal$boxHeight, \n",
       "            unitTo = get(\"page_units\", envir = pgEnv), valueOnly = TRUE)\n",
       "        spaceHeight <- boxHeight * (transcriptsInternal$spaceHeight)\n",
       "        textHeight <- convertHeight(textHeight, unitTo = get(\"page_units\", \n",
       "            envir = pgEnv), valueOnly = TRUE)\n",
       "        unit <- get(\"page_units\", envir = pgEnv)\n",
       "    }\n",
       "    maxRows <- floor(height/(boxHeight + spaceHeight + textHeight + \n",
       "        0.25 * textHeight))\n",
       "    wiggle <- abs(transcripts$chromend - transcripts$chromstart) * \n",
       "        transcriptsInternal$spaceWidth\n",
       "    if (transcriptsInternal$strandSplit == FALSE) {\n",
       "        repData <- data[duplicated(data$TXNAME) == FALSE, ]\n",
       "        repData <- defaultGenePriorities(data = repData, assembly = transcripts$assembly, \n",
       "            transcript = TRUE)\n",
       "        rowData <- assignRows(data = repData[, c(\"TXSTART\", \"TXEND\", \n",
       "            \"TXID\")], maxRows = maxRows, wiggle = wiggle, rowCol = 3, \n",
       "            limitLabel = transcriptsInternal$limitLabel, gTree = \"transcript_grobs\")\n",
       "        rowData <- suppressMessages(dplyr::left_join(x = data, \n",
       "            y = rowData[, c(\"row\", \"TXID\")], by = \"TXID\"))\n",
       "        rowData$y <- rowData$row * (boxHeight + spaceHeight + \n",
       "            textHeight + 0.25 * textHeight)\n",
       "    }\n",
       "    else {\n",
       "        repPos <- posStrand[duplicated(posStrand$TXNAME) == FALSE, \n",
       "            ]\n",
       "        repMin <- minStrand[duplicated(minStrand$TXNAME) == FALSE, \n",
       "            ]\n",
       "        repPos <- defaultGenePriorities(data = repPos, assembly = transcripts$assembly, \n",
       "            transcript = TRUE)\n",
       "        repMin <- defaultGenePriorities(data = repMin, assembly = transcripts$assembly, \n",
       "            transcript = TRUE)\n",
       "        posData <- assignRows(data = repPos[, c(\"TXSTART\", \"TXEND\", \n",
       "            \"TXID\")], maxRows = floor(maxRows/2), wiggle = wiggle, \n",
       "            rowCol = 3, limitLabel = transcriptsInternal$limitLabel, \n",
       "            gTree = \"transcript_grobs\")\n",
       "        minData <- assignRows(data = repMin[, c(\"TXSTART\", \"TXEND\", \n",
       "            \"TXID\")], maxRows = floor(maxRows/2), wiggle = wiggle, \n",
       "            rowCol = 3, limitLabel = transcriptsInternal$limitLabel, \n",
       "            gTree = \"transcript_grobs\", side = \"bottom\")\n",
       "        posData <- suppressMessages(dplyr::left_join(x = posStrand, \n",
       "            y = posData[, c(\"row\", \"TXID\")], by = \"TXID\"))\n",
       "        minData <- suppressMessages(dplyr::left_join(x = minStrand, \n",
       "            y = minData[, c(\"row\", \"TXID\")], by = \"TXID\"))\n",
       "        posData$y <- (0.5 * spaceHeight) + posData$row * (boxHeight + \n",
       "            spaceHeight + textHeight + 0.25 * textHeight)\n",
       "        minData$y <- ((0.5 * spaceHeight + boxHeight + textHeight + \n",
       "            0.25 * textHeight) + minData$row * (boxHeight + spaceHeight + \n",
       "            textHeight + 0.25 * textHeight)) * -1\n",
       "        rowData <- rbind(posData, minData)\n",
       "    }\n",
       "    if (nrow(rowData) > 0) {\n",
       "        if ((transcripts$chromend - transcripts$chromstart) >= \n",
       "            2.5e+07) {\n",
       "            transcriptLine <- rectGrob(x = rowData$TXSTART, y = rowData$y, \n",
       "                width = rowData$TXEND - rowData$TXSTART, height = boxHeight, \n",
       "                just = c(\"left\", \"bottom\"), gp = gpar(fill = rowData$color, \n",
       "                  col = rowData$color, lwd = transcriptsInternal$stroke), \n",
       "                default.units = \"native\")\n",
       "        }\n",
       "        else {\n",
       "            transcriptLine <- rectGrob(x = rowData$TXSTART, y = rowData$y + \n",
       "                0.5 * boxHeight, width = rowData$TXEND - rowData$TXSTART, \n",
       "                height = boxHeight * 0.2, just = \"left\", gp = gpar(fill = rowData$color, \n",
       "                  col = rowData$color, lwd = transcriptsInternal$stroke), \n",
       "                default.units = \"native\")\n",
       "            transcriptExons <- rectGrob(x = rowData$EXONSTART, \n",
       "                y = rowData$y + 0.5 * boxHeight, width = rowData$EXONEND - \n",
       "                  rowData$EXONSTART, height = boxHeight * 0.65, \n",
       "                just = \"left\", gp = gpar(fill = rowData$color, \n",
       "                  col = NA), default.units = \"native\")\n",
       "            assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", \n",
       "                envir = pgEnv), child = transcriptExons), envir = pgEnv)\n",
       "            cdsData <- rowData[which(!is.na(rowData$CDSSTART)), \n",
       "                ]\n",
       "            if (nrow(cdsData) > 0) {\n",
       "                transcriptCds <- rectGrob(x = cdsData$CDSSTART, \n",
       "                  y = cdsData$y + 0.5 * boxHeight, width = cdsData$CDSEND - \n",
       "                    cdsData$CDSSTART, height = boxHeight, just = \"left\", \n",
       "                  gp = gpar(fill = cdsData$color, col = cdsData$color, \n",
       "                    lwd = 1.25), default.units = \"native\")\n",
       "                assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", \n",
       "                  envir = pgEnv), child = transcriptCds), envir = pgEnv)\n",
       "            }\n",
       "        }\n",
       "        assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", \n",
       "            envir = pgEnv), child = transcriptLine), envir = pgEnv)\n",
       "        if (!is.null(transcriptsInternal$labels)) {\n",
       "            transcriptNames <- rowData[duplicated(rowData$TXNAME) == \n",
       "                FALSE, ]\n",
       "            transcriptNames$labelLoc <- rowMeans(transcriptNames[c(\"TXSTART\", \n",
       "                \"TXEND\")])\n",
       "            if (transcriptsInternal$labels == \"transcript\") {\n",
       "                label <- transcriptNames$TXNAME\n",
       "            }\n",
       "            else if (transcriptsInternal$labels == \"gene\") {\n",
       "                label <- transcriptNames[[transcripts$assembly$display.column]]\n",
       "            }\n",
       "            else {\n",
       "                label <- paste0(transcriptNames[[transcripts$assembly$display.column]], \n",
       "                  \":\", transcriptNames$TXNAME)\n",
       "            }\n",
       "            checkedLabels <- apply(data.frame(label = label, \n",
       "                labelLoc = transcriptNames$labelLoc), 1, cutoffLabel, \n",
       "                fontsize = transcriptsInternal$fontsize, xscale = transcriptsInternal$xscale, \n",
       "                vp = vp, unit = unit)\n",
       "            transcriptNames$label <- checkedLabels\n",
       "            transcriptNames <- transcriptNames[!is.na(transcriptNames$label), \n",
       "                ]\n",
       "            if (nrow(transcriptNames) > 0) {\n",
       "                transcript_names <- textGrob(label = transcriptNames$label, \n",
       "                  x = transcriptNames$labelLoc, y = transcriptNames$y + \n",
       "                    boxHeight + textHeight * 0.25, just = \"bottom\", \n",
       "                  gp = gpar(col = transcriptNames$color, fontsize = transcriptsInternal$fontsize), \n",
       "                  default.units = \"native\", check.overlap = TRUE)\n",
       "                assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", \n",
       "                  envir = pgEnv), child = transcript_names), \n",
       "                  envir = pgEnv)\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "    if (transcriptsInternal$draw == TRUE) {\n",
       "        grid.draw(get(\"transcript_grobs\", envir = pgEnv))\n",
       "    }\n",
       "    transcripts$grobs <- get(\"transcript_grobs\", envir = pgEnv)\n",
       "    message(\"transcripts[\", vp$name, \"]\")\n",
       "    invisible(transcripts)\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function (chrom, chromstart = NULL, chromend = NULL, assembly = \"hg38\", \n",
       "    fill = c(\"#669fd9\", \"#abcc8e\"), colorbyStrand = TRUE, strandSplit = FALSE, \n",
       "    boxHeight = unit(2, \"mm\"), spaceWidth = 0.02, spaceHeight = 0.3, \n",
       "    limitLabel = TRUE, fontsize = 8, labels = \"transcript\", stroke = 0.1, \n",
       "    bg = NA, x = NULL, y = NULL, width = NULL, height = NULL, \n",
       "    just = c(\"left\", \"top\"), default.units = \"inches\", draw = TRUE, \n",
       "    params = NULL) \n",
       "{\n",
       "    errorcheck_plotTranscripts <- function(transcriptPlot, labels, \n",
       "        fill) {\n",
       "        regionErrors(chromstart = transcriptPlot$chromstart, \n",
       "            chromend = transcriptPlot$chromend)\n",
       "        if (!labels %in% c(NULL, \"transcript\", \"gene\", \"both\")) {\n",
       "            stop(\"Invalid 'labels' input. Options are 'NULL', \", \n",
       "                \"'transcript', 'gene', or 'both'.\", call. = FALSE)\n",
       "        }\n",
       "        checkColorby(fill = fill, colorby = FALSE)\n",
       "    }\n",
       "    strand_scale <- function(strandSplit, height) {\n",
       "        if (strandSplit == TRUE) {\n",
       "            yscale <- c(-height/2, height/2)\n",
       "        }\n",
       "        else {\n",
       "            yscale <- c(0, height)\n",
       "        }\n",
       "        return(yscale)\n",
       "    }\n",
       "    transcriptsInternal <- parseParams(params = params, defaultArgs = formals(eval(match.call()[[1]])), \n",
       "        declaredArgs = lapply(match.call()[-1], eval.parent, \n",
       "            n = 2), class = \"transcriptsInternal\")\n",
       "    transcriptsInternal$just <- justConversion(just = transcriptsInternal$just)\n",
       "    transcripts <- structure(list(chrom = transcriptsInternal$chrom, \n",
       "        chromstart = transcriptsInternal$chromstart, chromend = transcriptsInternal$chromend, \n",
       "        assembly = transcriptsInternal$assembly, x = transcriptsInternal$x, \n",
       "        y = transcriptsInternal$y, width = transcriptsInternal$width, \n",
       "        height = transcriptsInternal$height, just = transcriptsInternal$just, \n",
       "        grobs = NULL), class = \"transcripts\")\n",
       "    attr(x = transcripts, which = \"plotted\") <- transcriptsInternal$draw\n",
       "    if (is.null(transcripts$chrom)) {\n",
       "        stop(\"argument \\\"chrom\\\" is missing, with no default.\", \n",
       "            call. = FALSE)\n",
       "    }\n",
       "    check_placement(object = transcripts)\n",
       "    errorcheck_plotTranscripts(transcriptPlot = transcripts, \n",
       "        labels = transcriptsInternal$labels, fill = transcriptsInternal$fill)\n",
       "    transcripts$assembly <- parseAssembly(assembly = transcripts$assembly)\n",
       "    transcripts <- defaultUnits(object = transcripts, default.units = transcriptsInternal$default.units)\n",
       "    transcriptsInternal$boxHeight <- misc_defaultUnits(value = transcriptsInternal$boxHeight, \n",
       "        name = \"boxHeight\", default.units = transcriptsInternal$default.units)\n",
       "    buildData <- geneData(object = transcripts, objectInternal = transcriptsInternal)\n",
       "    transcripts <- buildData[[1]]\n",
       "    transcriptsInternal <- buildData[[2]]\n",
       "    data <- transcriptsInternal$data\n",
       "    data$length <- data$TXEND - data$TXSTART\n",
       "    if (transcriptsInternal$colorbyStrand == TRUE) {\n",
       "        if (length(transcriptsInternal$fill) == 1) {\n",
       "            posCol <- transcriptsInternal$fill\n",
       "            negCol <- transcriptsInternal$fill\n",
       "        }\n",
       "        else {\n",
       "            posCol <- transcriptsInternal$fill[1]\n",
       "            negCol <- transcriptsInternal$fill[2]\n",
       "        }\n",
       "        pos <- data[which(data$TXSTRAND == \"+\"), ]\n",
       "        pos$color <- rep(posCol, nrow(pos))\n",
       "        neg <- data[which(data$TXSTRAND == \"-\"), ]\n",
       "        neg$color <- rep(negCol, nrow(neg))\n",
       "        data <- rbind(pos, neg)\n",
       "    }\n",
       "    else {\n",
       "        data$color <- rep(transcriptsInternal$fill[1], nrow(data))\n",
       "    }\n",
       "    if (transcriptsInternal$strandSplit == TRUE) {\n",
       "        posStrand <- data[which(data$TXSTRAND == \"+\"), ]\n",
       "        minStrand <- data[which(data$TXSTRAND == \"-\"), ]\n",
       "    }\n",
       "    currentViewports <- current_viewports()\n",
       "    vp_name <- paste0(\"transcripts\", length(grep(pattern = \"transcripts\", \n",
       "        x = currentViewports)) + 1)\n",
       "    if (is.null(transcripts$x) | is.null(transcripts$y)) {\n",
       "        height <- 0.5\n",
       "        yscale <- strand_scale(strandSplit = transcriptsInternal$strandSplit, \n",
       "            height = height)\n",
       "        vp <- viewport(height = unit(0.5, \"snpc\"), width = unit(1, \n",
       "            \"snpc\"), x = unit(0.5, \"npc\"), y = unit(0.5, \"npc\"), \n",
       "            clip = \"on\", xscale = transcriptsInternal$xscale, \n",
       "            yscale = yscale, just = \"center\", name = vp_name)\n",
       "        if (transcriptsInternal$draw == TRUE) {\n",
       "            vp$name <- \"transcripts1\"\n",
       "            grid.newpage()\n",
       "        }\n",
       "    }\n",
       "    else {\n",
       "        addViewport(vp_name)\n",
       "        page_coords <- convert_page(object = transcripts)\n",
       "        height <- convertHeight(page_coords$height, unitTo = get(\"page_units\", \n",
       "            envir = pgEnv), valueOnly = TRUE)\n",
       "        yscale <- strand_scale(strandSplit = transcriptsInternal$strandSplit, \n",
       "            height = height)\n",
       "        vp <- viewport(height = page_coords$height, width = page_coords$width, \n",
       "            x = page_coords$x, y = page_coords$y, clip = \"on\", \n",
       "            xscale = transcriptsInternal$xscale, yscale = yscale, \n",
       "            just = transcriptsInternal$just, name = vp_name)\n",
       "    }\n",
       "    backgroundGrob <- rectGrob(gp = gpar(fill = transcriptsInternal$bg, \n",
       "        col = NA), name = \"background\")\n",
       "    assign(\"transcript_grobs\", gTree(vp = vp, children = gList(backgroundGrob)), \n",
       "        envir = pgEnv)\n",
       "    if (is.null(transcriptsInternal$labels)) {\n",
       "        textHeight <- unit(0, \"npc\")\n",
       "    }\n",
       "    else {\n",
       "        textHeight <- heightDetails(textGrob(label = \"A\", gp = gpar(fontsize = transcriptsInternal$fontsize)))\n",
       "    }\n",
       "    if (is.null(transcripts$x) & is.null(transcripts$y)) {\n",
       "        pushViewport(vp)\n",
       "        boxHeight <- convertHeight(transcriptsInternal$boxHeight, \n",
       "            unitTo = \"npc\", valueOnly = TRUE)\n",
       "        spaceHeight <- boxHeight * (transcriptsInternal$spaceHeight)\n",
       "        textHeight <- convertHeight(textHeight, unitTo = \"npc\", \n",
       "            valueOnly = TRUE)\n",
       "        upViewport()\n",
       "        unit <- \"npc\"\n",
       "    }\n",
       "    else {\n",
       "        boxHeight <- convertHeight(transcriptsInternal$boxHeight, \n",
       "            unitTo = get(\"page_units\", envir = pgEnv), valueOnly = TRUE)\n",
       "        spaceHeight <- boxHeight * (transcriptsInternal$spaceHeight)\n",
       "        textHeight <- convertHeight(textHeight, unitTo = get(\"page_units\", \n",
       "            envir = pgEnv), valueOnly = TRUE)\n",
       "        unit <- get(\"page_units\", envir = pgEnv)\n",
       "    }\n",
       "    maxRows <- floor(height/(boxHeight + spaceHeight + textHeight + \n",
       "        0.25 * textHeight))\n",
       "    wiggle <- abs(transcripts$chromend - transcripts$chromstart) * \n",
       "        transcriptsInternal$spaceWidth\n",
       "    if (transcriptsInternal$strandSplit == FALSE) {\n",
       "        repData <- data[duplicated(data$TXNAME) == FALSE, ]\n",
       "        repData <- defaultGenePriorities(data = repData, assembly = transcripts$assembly, \n",
       "            transcript = TRUE)\n",
       "        rowData <- assignRows(data = repData[, c(\"TXSTART\", \"TXEND\", \n",
       "            \"TXID\")], maxRows = maxRows, wiggle = wiggle, rowCol = 3, \n",
       "            limitLabel = transcriptsInternal$limitLabel, gTree = \"transcript_grobs\")\n",
       "        rowData <- suppressMessages(dplyr::left_join(x = data, \n",
       "            y = rowData[, c(\"row\", \"TXID\")], by = \"TXID\"))\n",
       "        rowData$y <- rowData$row * (boxHeight + spaceHeight + \n",
       "            textHeight + 0.25 * textHeight)\n",
       "    }\n",
       "    else {\n",
       "        repPos <- posStrand[duplicated(posStrand$TXNAME) == FALSE, \n",
       "            ]\n",
       "        repMin <- minStrand[duplicated(minStrand$TXNAME) == FALSE, \n",
       "            ]\n",
       "        repPos <- defaultGenePriorities(data = repPos, assembly = transcripts$assembly, \n",
       "            transcript = TRUE)\n",
       "        repMin <- defaultGenePriorities(data = repMin, assembly = transcripts$assembly, \n",
       "            transcript = TRUE)\n",
       "        posData <- assignRows(data = repPos[, c(\"TXSTART\", \"TXEND\", \n",
       "            \"TXID\")], maxRows = floor(maxRows/2), wiggle = wiggle, \n",
       "            rowCol = 3, limitLabel = transcriptsInternal$limitLabel, \n",
       "            gTree = \"transcript_grobs\")\n",
       "        minData <- assignRows(data = repMin[, c(\"TXSTART\", \"TXEND\", \n",
       "            \"TXID\")], maxRows = floor(maxRows/2), wiggle = wiggle, \n",
       "            rowCol = 3, limitLabel = transcriptsInternal$limitLabel, \n",
       "            gTree = \"transcript_grobs\", side = \"bottom\")\n",
       "        posData <- suppressMessages(dplyr::left_join(x = posStrand, \n",
       "            y = posData[, c(\"row\", \"TXID\")], by = \"TXID\"))\n",
       "        minData <- suppressMessages(dplyr::left_join(x = minStrand, \n",
       "            y = minData[, c(\"row\", \"TXID\")], by = \"TXID\"))\n",
       "        posData$y <- (0.5 * spaceHeight) + posData$row * (boxHeight + \n",
       "            spaceHeight + textHeight + 0.25 * textHeight)\n",
       "        minData$y <- ((0.5 * spaceHeight + boxHeight + textHeight + \n",
       "            0.25 * textHeight) + minData$row * (boxHeight + spaceHeight + \n",
       "            textHeight + 0.25 * textHeight)) * -1\n",
       "        rowData <- rbind(posData, minData)\n",
       "    }\n",
       "    if (nrow(rowData) > 0) {\n",
       "        if ((transcripts$chromend - transcripts$chromstart) >= \n",
       "            2.5e+07) {\n",
       "            transcriptLine <- rectGrob(x = rowData$TXSTART, y = rowData$y, \n",
       "                width = rowData$TXEND - rowData$TXSTART, height = boxHeight, \n",
       "                just = c(\"left\", \"bottom\"), gp = gpar(fill = rowData$color, \n",
       "                  col = rowData$color, lwd = transcriptsInternal$stroke), \n",
       "                default.units = \"native\")\n",
       "        }\n",
       "        else {\n",
       "            transcriptLine <- rectGrob(x = rowData$TXSTART, y = rowData$y + \n",
       "                0.5 * boxHeight, width = rowData$TXEND - rowData$TXSTART, \n",
       "                height = boxHeight * 0.2, just = \"left\", gp = gpar(fill = rowData$color, \n",
       "                  col = rowData$color, lwd = transcriptsInternal$stroke), \n",
       "                default.units = \"native\")\n",
       "            transcriptExons <- rectGrob(x = rowData$EXONSTART, \n",
       "                y = rowData$y + 0.5 * boxHeight, width = rowData$EXONEND - \n",
       "                  rowData$EXONSTART, height = boxHeight * 0.65, \n",
       "                just = \"left\", gp = gpar(fill = rowData$color, \n",
       "                  col = NA), default.units = \"native\")\n",
       "            assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", \n",
       "                envir = pgEnv), child = transcriptExons), envir = pgEnv)\n",
       "            cdsData <- rowData[which(!is.na(rowData$CDSSTART)), \n",
       "                ]\n",
       "            if (nrow(cdsData) > 0) {\n",
       "                transcriptCds <- rectGrob(x = cdsData$CDSSTART, \n",
       "                  y = cdsData$y + 0.5 * boxHeight, width = cdsData$CDSEND - \n",
       "                    cdsData$CDSSTART, height = boxHeight, just = \"left\", \n",
       "                  gp = gpar(fill = cdsData$color, col = cdsData$color, \n",
       "                    lwd = 1.25), default.units = \"native\")\n",
       "                assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", \n",
       "                  envir = pgEnv), child = transcriptCds), envir = pgEnv)\n",
       "            }\n",
       "        }\n",
       "        assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", \n",
       "            envir = pgEnv), child = transcriptLine), envir = pgEnv)\n",
       "        if (!is.null(transcriptsInternal$labels)) {\n",
       "            transcriptNames <- rowData[duplicated(rowData$TXNAME) == \n",
       "                FALSE, ]\n",
       "            transcriptNames$labelLoc <- rowMeans(transcriptNames[c(\"TXSTART\", \n",
       "                \"TXEND\")])\n",
       "            if (transcriptsInternal$labels == \"transcript\") {\n",
       "                label <- transcriptNames$TXNAME\n",
       "            }\n",
       "            else if (transcriptsInternal$labels == \"gene\") {\n",
       "                label <- transcriptNames[[transcripts$assembly$display.column]]\n",
       "            }\n",
       "            else {\n",
       "                label <- paste0(transcriptNames[[transcripts$assembly$display.column]], \n",
       "                  \":\", transcriptNames$TXNAME)\n",
       "            }\n",
       "            checkedLabels <- apply(data.frame(label = label, \n",
       "                labelLoc = transcriptNames$labelLoc), 1, cutoffLabel, \n",
       "                fontsize = transcriptsInternal$fontsize, xscale = transcriptsInternal$xscale, \n",
       "                vp = vp, unit = unit)\n",
       "            transcriptNames$label <- checkedLabels\n",
       "            transcriptNames <- transcriptNames[!is.na(transcriptNames$label), \n",
       "                ]\n",
       "            if (nrow(transcriptNames) > 0) {\n",
       "                transcript_names <- textGrob(label = transcriptNames$label, \n",
       "                  x = transcriptNames$labelLoc, y = transcriptNames$y + \n",
       "                    boxHeight + textHeight * 0.25, just = \"bottom\", \n",
       "                  gp = gpar(col = transcriptNames$color, fontsize = transcriptsInternal$fontsize), \n",
       "                  default.units = \"native\", check.overlap = TRUE)\n",
       "                assign(\"transcript_grobs\", addGrob(get(\"transcript_grobs\", \n",
       "                  envir = pgEnv), child = transcript_names), \n",
       "                  envir = pgEnv)\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "    if (transcriptsInternal$draw == TRUE) {\n",
       "        grid.draw(get(\"transcript_grobs\", envir = pgEnv))\n",
       "    }\n",
       "    transcripts$grobs <- get(\"transcript_grobs\", envir = pgEnv)\n",
       "    message(\"transcripts[\", vp$name, \"]\")\n",
       "    invisible(transcripts)\n",
       "}\n",
       "<bytecode: 0x561131582860>\n",
       "<environment: namespace:plotgardener>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotTranscripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e8b2d75",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
